// @progress/kendo-angular-upload/esm2022/upload.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_UPLOAD } from './directives';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { PopupService } from '@progress/kendo-angular-popup';
import * as i0 from "@angular/core";
import * as i1 from "./upload.component";
import * as i2 from "./rendering/upload-action-buttons.component";
import * as i3 from "./rendering/upload-status-total.component";
import * as i4 from "./templates/file-template.directive";
import * as i5 from "./templates/file-info-template.directive";
import * as i6 from "./localization/custom-messages.component";
import * as i7 from "./dropzone-external.directive";
import * as i8 from "./dropzone.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](https://angular.io/api/core/NgModule) definition for the Upload component.
 */
export class UploadModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: UploadModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: UploadModule, imports: [i1.UploadComponent, i2.UploadActionButtonsComponent, i3.UploadStatusTotalComponent, i4.FileTemplateDirective, i5.FileInfoTemplateDirective, i6.CustomMessagesComponent, i7.UploadDropZoneDirective, i8.UploadDropZoneComponent], exports: [i1.UploadComponent, i2.UploadActionButtonsComponent, i3.UploadStatusTotalComponent, i4.FileTemplateDirective, i5.FileInfoTemplateDirective, i6.CustomMessagesComponent, i7.UploadDropZoneDirective, i8.UploadDropZoneComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: UploadModule, providers: [
            IconsService,
            PopupService,
            ResizeBatchService
        ], imports: [i1.UploadComponent, i3.UploadStatusTotalComponent, i8.UploadDropZoneComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: UploadModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_UPLOAD],
                    imports: [...KENDO_UPLOAD],
                    providers: [
                        IconsService,
                        PopupService,
                        ResizeBatchService
                    ]
                }]
        }] });
