// @progress/kendo-angular-pager/esm2022/pager/spacer.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, HostBinding, Input } from '@angular/core';
import { isPresent } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
/**
 * Represents the Kendo UI Pager Spacer component for Angular.
 * Gives you additional white space between the Pager inner elements,
 * and provides a way for customizing the spacer width.
 *
 * @example
 * ```html
 * <kendo-pager [skip]="skip" [pageSize]="pageSize" [total]="total">
 *   <ng-template kendoPagerTemplate>
 *     <kendo-pager-prev-buttons></kendo-pager-prev-buttons>
 *     <kendo-pager-info></kendo-pager-info>
 *     <kendo-pager-spacer></kendo-pager-spacer>
 *     <kendo-pager-next-buttons></kendo-pager-next-buttons>
 *   </ng-template>
 * </kendo-pager>
 * ```
 */
export class PagerSpacerComponent {
    /**
     * Gets the CSS class for the spacer.
     */
    hostClass = true;
    /**
     * Gets the CSS class for sized spacers.
     */
    get sizedClass() {
        return isPresent(this.width);
    }
    /**
     * Gets the flex-basis style for the spacer width.
     */
    get flexBasisStyle() {
        return this.width;
    }
    /**
     * Specifies the width of the `PagerSpacer` component.
     * Accepts the [string values of the CSS `flex-basis` property](https://developer.mozilla.org/en-US/docs/Web/CSS/flex-basis).
     *
     * If you do not set this property, the `PagerSpacer` takes all the available space.
     */
    width;
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: PagerSpacerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.14", type: PagerSpacerComponent, isStandalone: true, selector: "kendo-pager-spacer", inputs: { width: "width" }, host: { properties: { "class.k-spacer": "this.hostClass", "class.k-spacer-sized": "this.sizedClass", "style.flexBasis": "this.flexBasisStyle" } }, ngImport: i0, template: ``, isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: PagerSpacerComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'kendo-pager-spacer',
                    template: ``,
                    standalone: true
                }]
        }], propDecorators: { hostClass: [{
                type: HostBinding,
                args: ['class.k-spacer']
            }], sizedClass: [{
                type: HostBinding,
                args: ['class.k-spacer-sized']
            }], flexBasisStyle: [{
                type: HostBinding,
                args: ['style.flexBasis']
            }], width: [{
                type: Input
            }] } });
