// @progress/kendo-angular-dateinputs/esm2022/calendar/calendars.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_MULTIVIEWCALENDAR, KENDO_CALENDAR } from '../directives';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { CenturyViewService } from './services/century-view.service';
import { DecadeViewService } from './services/decade-view.service';
import { CalendarDOMService } from './services/dom.service';
import { MonthViewService } from './services/month-view.service';
import { NavigationService } from './services/navigation.service';
import { YearViewService } from './services/year-view.service';
import * as i0 from "@angular/core";
import * as i1 from "./localization/calendar-custom-messages.component";
import * as i2 from "./calendar.component";
import * as i3 from "./templates/cell-template.directive";
import * as i4 from "./templates/century-cell-template.directive";
import * as i5 from "./templates/decade-cell-template.directive";
import * as i6 from "./templates/footer-template.directive";
import * as i7 from "./templates/header-template.directive";
import * as i8 from "./templates/header-title-template.directive";
import * as i9 from "./templates/month-cell-template.directive";
import * as i10 from "./templates/navigation-item-template.directive";
import * as i11 from "./templates/weeknumber-cell-template.directive";
import * as i12 from "./templates/year-cell-template.directive";
import * as i13 from "./localization/multiview-calendar-custom-messages.component";
import * as i14 from "./multiview-calendar.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Required for adding all Calendars features in NgModule-based Angular applications.
 *
 * The package exports:
 * - `CalendarModule`&mdash;The calendar module.
 * - `MultiViewCalendarModule`&mdash;The multi-view calendar module.
 *
 * @example
 *
 * ```typescript
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { CalendarsModule } from '@progress/kendo-angular-dateinputs';
 *
 * @NgModule({
 *   imports: [
 *     BrowserModule,
 *     CalendarsModule
 *   ]
 * })
 * export class AppModule { }
 * ```
 */
export class CalendarsModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CalendarsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: CalendarsModule, imports: [i1.CalendarCustomMessagesComponent, i2.CalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective, i13.MultiViewCalendarCustomMessagesComponent, i14.MultiViewCalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective], exports: [i1.CalendarCustomMessagesComponent, i2.CalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective, i13.MultiViewCalendarCustomMessagesComponent, i14.MultiViewCalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CalendarsModule, providers: [
            IconsService,
            ResizeBatchService,
            CalendarDOMService,
            CenturyViewService,
            DecadeViewService,
            MonthViewService,
            YearViewService,
            PopupService,
            NavigationService
        ], imports: [i2.CalendarComponent, i14.MultiViewCalendarComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CalendarsModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_CALENDAR, ...KENDO_MULTIVIEWCALENDAR],
                    imports: [...KENDO_CALENDAR, ...KENDO_MULTIVIEWCALENDAR],
                    providers: [
                        IconsService,
                        ResizeBatchService,
                        CalendarDOMService,
                        CenturyViewService,
                        DecadeViewService,
                        MonthViewService,
                        YearViewService,
                        PopupService,
                        NavigationService
                    ]
                }]
        }] });
