// @progress/kendo-angular-dateinputs/esm2022/calendar/calendar.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright ôˋ 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { KENDO_CALENDAR } from "../directives";
import { CenturyViewService } from "./services/century-view.service";
import { DecadeViewService } from "./services/decade-view.service";
import { CalendarDOMService } from "./services/dom.service";
import { MonthViewService } from "./services/month-view.service";
import { YearViewService } from "./services/year-view.service";
import { IconsService } from "@progress/kendo-angular-icons";
import { PopupService } from "@progress/kendo-angular-popup";
import { NavigationService } from "./services/navigation.service";
import { ResizeBatchService } from "@progress/kendo-angular-common";
import * as i0 from "@angular/core";
import * as i1 from "./localization/calendar-custom-messages.component";
import * as i2 from "./calendar.component";
import * as i3 from "./templates/cell-template.directive";
import * as i4 from "./templates/century-cell-template.directive";
import * as i5 from "./templates/decade-cell-template.directive";
import * as i6 from "./templates/footer-template.directive";
import * as i7 from "./templates/header-template.directive";
import * as i8 from "./templates/header-title-template.directive";
import * as i9 from "./templates/month-cell-template.directive";
import * as i10 from "./templates/navigation-item-template.directive";
import * as i11 from "./templates/weeknumber-cell-template.directive";
import * as i12 from "./templates/year-cell-template.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Required for adding all Calendar features in NgModule-based Angular applications.
 *
 * The package exports:
 * - `CellTemplateDirective`&mdash;The month cell template directive.
 * - `MonthCellTemplateDirective`&mdash;The month cell template directive.
 * - `YearCellTemplateDirective`&mdash;The year cell template directive.
 * - `DecadeCellTemplateDirective`&mdash;The decade cell template directive.
 * - `CenturyCellTemplateDirective`&mdash;The century cell template directive.
 * - `WeekNumberCellTemplateDirective`&mdash;The month week number cell template directive.
 * - `HeaderTitleTemplateDirective`&mdash;The header title template directive.
 * - `NavigationItemTemplateDirective`&mdash;The navigation item template directive.
 *
 * @example
 *
 * ```typescript
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { CalendarModule } from '@progress/kendo-angular-dateinputs';
 *
 * @NgModule({
 *   imports: [
 *     BrowserModule,
 *     CalendarModule
 *   ]
 * })
 * export class AppModule { }
 * ```
 */
export class CalendarModule {
    static èçfac = i0.èçèçngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CalendarModule, deps: [], target: i0.èçèçFactoryTarget.NgModule });
    static èçmod = i0.èçèçngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: CalendarModule, imports: [i1.CalendarCustomMessagesComponent, i2.CalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective], exports: [i1.CalendarCustomMessagesComponent, i2.CalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective] });
    static èçinj = i0.èçèçngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CalendarModule, providers: [
            IconsService,
            ResizeBatchService,
            CalendarDOMService,
            CenturyViewService,
            DecadeViewService,
            MonthViewService,
            YearViewService,
            PopupService,
            NavigationService
        ], imports: [i2.CalendarComponent] });
}
i0.èçèçngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CalendarModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_CALENDAR],
                    imports: [...KENDO_CALENDAR],
                    providers: [
                        IconsService,
                        ResizeBatchService,
                        CalendarDOMService,
                        CenturyViewService,
                        DecadeViewService,
                        MonthViewService,
                        YearViewService,
                        PopupService,
                        NavigationService
                    ]
                }]
        }] });
