// @progress/kendo-angular-progressbar/esm2022/progressbar.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_PROGRESSBARS } from './directives';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./chunk/chunk-progressbar.component";
import * as i2 from "./common/localization/custom-messages.component";
import * as i3 from "./circularprogressbar/center-template.directive";
import * as i4 from "./circularprogressbar/circular-progressbar.component";
import * as i5 from "./progressbar.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmodules'])
 * definition for the ProgressBar components.
 *
 * @example
 * ```typescript
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { ProgressBarModule } from '@progress/kendo-angular-progressbar';
 *
 * @NgModule({
 *   imports: [BrowserModule, ProgressBarModule],
 *   declarations: [AppComponent],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule { }
 * ```
 */
export class ProgressBarModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ProgressBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: ProgressBarModule, imports: [i1.ChunkProgressBarComponent, i2.ProgressBarCustomMessagesComponent, i3.CircularProgressbarCenterTemplateDirective, i4.CircularProgressBarComponent, i2.ProgressBarCustomMessagesComponent, i5.ProgressBarComponent, i2.ProgressBarCustomMessagesComponent], exports: [i1.ChunkProgressBarComponent, i2.ProgressBarCustomMessagesComponent, i3.CircularProgressbarCenterTemplateDirective, i4.CircularProgressBarComponent, i2.ProgressBarCustomMessagesComponent, i5.ProgressBarComponent, i2.ProgressBarCustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ProgressBarModule, providers: [ResizeBatchService], imports: [i4.CircularProgressBarComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ProgressBarModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_PROGRESSBARS],
                    imports: [...KENDO_PROGRESSBARS],
                    providers: [ResizeBatchService]
                }]
        }] });
