// @progress/kendo-angular-indicators/esm2022/indicators.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_INDICATORS } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./badge/badge.component";
import * as i2 from "./badge/badge-container.component";
import * as i3 from "./loader/loader.component";
import * as i4 from "./skeleton/skeleton.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Indicators components.
 *
 * @example
 *
 * ```ts-no-run
 * import { IndicatorsModule } from '@progress/kendo-angular-indicators';
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { NgModule } from '@angular/core';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *   declarations: [AppComponent],
 *   imports: [BrowserModule, IndicatorsModule],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 *
 * platformBrowserDynamic().bootstrapModule(AppModule);
 * ```
 */
export class IndicatorsModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: IndicatorsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: IndicatorsModule, imports: [i1.BadgeComponent, i2.BadgeContainerComponent, i3.LoaderComponent, i4.SkeletonComponent], exports: [i1.BadgeComponent, i2.BadgeContainerComponent, i3.LoaderComponent, i4.SkeletonComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: IndicatorsModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: IndicatorsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_INDICATORS],
                    exports: [...KENDO_INDICATORS]
                }]
        }] });
