// @progress/kendo-angular-dateinputs/esm2022/calendar/templates/cell-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * A directive for customizing the cells in the Calendar component.
 *
 * This directive allows you to define a custom template for rendering the content of calendar cells
 * in the month view. The template context provides access to the current date and cell context information.
 *
 * Template Context Variables:
 * - `date` - The Date object representing the current cell's date
 * - `cellContext` - Additional context information about the current cell
 *
 * For more examples, refer to the article on [templates]({% slug templates_calendar %}).
 *
 * `kendoCalendarCellTemplate` is equivalent to
 * [`kendoCalendarMonthCellTemplate`]({% slug api_dateinputs_monthcelltemplatedirective %}).
 *
 * @example
 * ```html
 * <kendo-calendar>
 *   <ng-template kendoCalendarCellTemplate let-date let-cellContext="cellContext">
 *     <span class="custom">{{date.getDate()}}</span>
 *   </ng-template>
 * </kendo-calendar>
 * ```
 */
export class CellTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CellTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: CellTemplateDirective, isStandalone: true, selector: "[kendoCalendarCellTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CellTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoCalendarCellTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
