// @progress/kendo-angular-dateinputs/esm2022/calendar/templates/footer-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * A directive for customizing the footer in the Calendar component.
 *
 * This directive allows you to define a custom template for rendering the footer content.
 * The template context provides access to the current date and active view.
 *
 * Template Context Variables:
 * - `date` - The Date object representing the current calendar context
 * - `activeView` - The current active view of the Calendar
 *
 * For more examples, refer to the article on [templates]({% slug templates_calendar %}).
 *
 * @example
 * ```html
 * <kendo-calendar>
 *   <ng-template kendoCalendarFooterTemplate let-date let-activeView="activeView">
 *     <div class="custom">
 *       Today: {{date | date:'short'}} ({{activeView}} view)
 *     </div>
 *   </ng-template>
 * </kendo-calendar>
 * ```
 */
export class FooterTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: FooterTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: FooterTemplateDirective, isStandalone: true, selector: "[kendoCalendarFooterTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: FooterTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoCalendarFooterTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
