// @progress/kendo-angular-dateinputs/esm2022/calendar/templates/header-title-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * A directive for customizing the header title in the Calendar component.
 *
 * This directive allows you to define a custom template for rendering the header title content.
 * The template context provides access to the title text, current date, and active view.
 *
 * Template Context Variables:
 * - `title` - The formatted title text for the header
 * - `date` - The Date object representing the current header context
 * - `activeView` - The current active view of the Calendar
 *
 * For more examples, refer to the article on [templates]({% slug templates_calendar %}).
 *
 * @example
 * ```html
 * <kendo-calendar>
 *   <ng-template kendoCalendarHeaderTitleTemplate let-title let-activeView="activeView">
 *     <span class="custom">{{title}} ({{activeView}})</span>
 *   </ng-template>
 * </kendo-calendar>
 * ```
 */
export class HeaderTitleTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: HeaderTitleTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: HeaderTitleTemplateDirective, isStandalone: true, selector: "[kendoCalendarHeaderTitleTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: HeaderTitleTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoCalendarHeaderTitleTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
