// @progress/kendo-angular-dateinputs/esm2022/calendar/templates/weeknumber-cell-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * A directive for customizing the week number cells in the Calendar component.
 *
 * This directive allows you to define a custom template for rendering the content of week number cells
 * in the month view when the weekNumber property is enabled. The template context provides access to
 * the current date and cell context information.
 *
 * Template Context Variables:
 * - `date` - The Date object representing the current week's date
 * - `cellContext` - Additional context information about the current week number cell
 *
 * For more examples, refer to the article on [templates]({% slug templates_calendar %}).
 *
 * @example
 * ```html
 * <kendo-calendar [weekNumber]="true">
 *   <ng-template kendoCalendarWeekNumberCellTemplate let-context="cellContext">
 *     <span class="custom">W{{context.formattedValue}}</span>
 *   </ng-template>
 * </kendo-calendar>
 * ```
 */
export class WeekNumberCellTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: WeekNumberCellTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: WeekNumberCellTemplateDirective, isStandalone: true, selector: "[kendoCalendarWeekNumberCellTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: WeekNumberCellTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoCalendarWeekNumberCellTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
