// @progress/kendo-angular-layout/esm2022/panelbar/panelbar-content.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Defines the content template of the declaratively initialized PanelBar items.
 * To define the content template, nest an `<ng-template>` tag
 * with the `kendoPanelBarContent` directive inside the `<kendo-panelbar-item>` tag.
 * The content can be expanded or collapsed through the item.
 *
 * @example
 * ```html
 * <kendo-panelbar-item title="Item">
 *   <div kendoPanelBarContent>
 *     <p>Item content goes here</p>
 *   </div>
 * </kendo-panelbar-item>
 * ```
 */
export class PanelBarContentDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: PanelBarContentDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: PanelBarContentDirective, isStandalone: true, selector: "[kendoPanelBarContent]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: PanelBarContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: "[kendoPanelBarContent]",
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
