// @progress/kendo-angular-layout/esm2022/timeline/templates/timeline-card-header.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Optional, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Specifies the header of each timeline card. To define a header template, nest an `<ng-template>` tag
 * with the `kendoTimelineCardHeaderTemplate` directive inside the `<kendo-timeline>` tag.
 *
 * The following values are available as context variables:
 * - `let-event="event"` (`TimelineEvent`)&mdash;The current card's event. Also available as implicit context variable.
 * - `let-index="index"` (`number`)&mdash;The current event index.
 *
 * @example
 * ```html
 * <kendo-timeline [events]="events">
 *   <ng-template kendoTimelineCardHeaderTemplate let-event let-index="index">
 *     <span class="k-event-title">Title: {{ event.title }}</span>
 *     <div kendoCardSubtitle>Event number {{ index }}</div>
 *   </ng-template>
 * </kendo-timeline>
 * ```
 */
export class TimelineCardHeaderTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TimelineCardHeaderTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: TimelineCardHeaderTemplateDirective, isStandalone: true, selector: "[kendoTimelineCardHeaderTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TimelineCardHeaderTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoTimelineCardHeaderTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }] });
