// @progress/kendo-angular-layout/esm2022/timeline/templates/timeline-card-body.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Optional, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Specifies the body of each timeline card. To define a body template, nest an `<ng-template>` tag
 * with the `kendoTimelineCardBodyTemplate` directive inside the `<kendo-timeline>` tag.
 *
 * The following values are available as context variables:
 * - `let-event="event"` (`TimelineEvent`)&mdash;The current card's event. Also available as implicit context variable.
 * - `let-index="index"` (`number`)&mdash;The current event index.
 *
 * @example
 * ```html
 * <kendo-timeline [events]="events">
 *   <ng-template kendoTimelineCardBodyTemplate let-event let-index="index">
 *     Component for the body of the card with index {{ index }} and event {{ event.title }}.
 *   </ng-template>
 * </kendo-timeline>
 * ```
 */
export class TimelineCardBodyTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TimelineCardBodyTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: TimelineCardBodyTemplateDirective, isStandalone: true, selector: "[kendoTimelineCardBodyTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TimelineCardBodyTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoTimelineCardBodyTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }] });
