// @progress/kendo-angular-grid/esm2022/filtering/string-filter.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
// eslint-disable no-access-missing-member
import { Directive, Input } from '@angular/core';
import { ColumnComponent } from '../columns/column.component';
import { FilterService } from './filter.service';
import { BaseFilterCellComponent, localizeOperators } from './base-filter-cell.component';
import { toJSON } from './operators/filter-operator.base';
import { ContextService } from '../common/provider.service';
import * as i0 from "@angular/core";
import * as i1 from "./filter.service";
import * as i2 from "../common/provider.service";
const stringOperators = localizeOperators({
    "filterEqOperator": "eq",
    "filterNotEqOperator": "neq",
    "filterContainsOperator": "contains",
    "filterNotContainsOperator": "doesnotcontain",
    "filterStartsWithOperator": "startswith",
    "filterEndsWithOperator": "endswith",
    "filterIsNullOperator": "isnull",
    "filterIsNotNullOperator": "isnotnull",
    "filterIsEmptyOperator": "isempty",
    "filterIsNotEmptyOperator": "isnotempty"
});
/**
 * @hidden
 * Represents a base string filter component.
 */
export class StringFilterComponent extends BaseFilterCellComponent {
    ctx;
    /**
     * Sets the column for this filter.
     * @type {ColumnComponent}
     */
    column;
    /**
     * Sets the default filter operator.
     * @type {string}
     * @default 'contains'
     */
    operator = "contains";
    /**
     * Gets the current filter for the column field.
     * @readonly
     * @type {FilterDescriptor}
     */
    get currentFilter() {
        return this.filterByField((this.column || {}).field);
    }
    /**
     * Gets the current filter operator for the column field.
     * @readonly
     * @type {string}
     */
    get currentOperator() {
        return this.currentFilter ? this.currentFilter.operator : this.operator;
    }
    subscription;
    constructor(filterService, ctx) {
        super(filterService);
        this.ctx = ctx;
    }
    ngOnInit() {
        this.subscription = this.ctx.localization.changes.subscribe(this.localizationChange.bind(this));
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
        super.ngOnDestroy();
    }
    localizationChange() {
        this.defaultOperators = stringOperators(this.ctx.localization);
        if (this.operatorList.length) {
            this.operators = toJSON(this.operatorList.toArray());
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: StringFilterComponent, deps: [{ token: i1.FilterService }, { token: i2.ContextService }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: StringFilterComponent, inputs: { column: "column", operator: "operator" }, usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: StringFilterComponent, decorators: [{
            type: Directive,
            args: [{}]
        }], ctorParameters: () => [{ type: i1.FilterService }, { type: i2.ContextService }], propDecorators: { column: [{
                type: Input
            }], operator: [{
                type: Input
            }] } });
