// @progress/kendo-angular-pdf-export/esm2022/pdf-export-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Optional, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Represents the directive for defining a custom page template in the PDF Export component ([see example](slug:multipagecontent_pdfexport#page-templates)).
 *
 * @example
 * ```html
 * <kendo-pdf-export paperSize="A4" margin="2cm">
 *    <ng-template kendoPDFTemplate let-pageNum="pageNum">
 *        <div>Page {{ pageNum }}</div>
 *    </ng-template>
 * </kendo-pdf-export>
 * ```
 */
export class PDFExportTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: PDFExportTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: PDFExportTemplateDirective, isStandalone: true, selector: "[kendoPDFTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: PDFExportTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoPDFTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }] });
