// @progress/kendo-angular-conversational-ui/esm2022/conversational-ui.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_CONVERSATIONALUI } from './directives';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { DialogContainerService, DialogService, WindowContainerService, WindowService } from '@progress/kendo-angular-dialog';
import * as i0 from "@angular/core";
import * as i1 from "./ai-prompt/aiprompt.component";
import * as i2 from "./ai-prompt/views/prompt-view.component";
import * as i3 from "./ai-prompt/views/output-view.component";
import * as i4 from "./ai-prompt/views/command-view.component";
import * as i5 from "./ai-prompt/views/custom-view.component";
import * as i6 from "./ai-prompt/localization/custom-messages.component";
import * as i7 from "./ai-prompt/templates/toolbar-actions.template";
import * as i8 from "./ai-prompt/common/toolbar-focusable.directive";
import * as i9 from "./ai-prompt/templates/aiprompt-output-template.directive";
import * as i10 from "./ai-prompt/templates/aiprompt-output-body-template.directive";
import * as i11 from "./chat/chat.component";
import * as i12 from "./chat/l10n/custom-messages.component";
import * as i13 from "./chat/templates/attachment-template.directive";
import * as i14 from "./chat/templates/author-message-content-template.directive";
import * as i15 from "./chat/templates/receiver-message-content-template.directive";
import * as i16 from "./chat/templates/message-content-template.directive";
import * as i17 from "./chat/templates/author-message-template.directive";
import * as i18 from "./chat/templates/receiver-message-template.directive";
import * as i19 from "./chat/templates/message-template.directive";
import * as i20 from "./chat/cards/hero-card.component";
import * as i21 from "./chat/templates/message-box.directive";
import * as i22 from "./chat/templates/header-template.directive";
import * as i23 from "./chat/templates/no-data-template.directive";
import * as i24 from "./chat/templates/timestamp-template.directive";
import * as i25 from "./chat/templates/status-template.directive";
import * as i26 from "./chat/templates/suggestion-template.directive";
import * as i27 from "./chat/templates/user-status-template.directive";
import * as i28 from "./inline-ai-prompt/inlineaiprompt.component";
import * as i29 from "./inline-ai-prompt/output-template.directive";
import * as i30 from "./inline-ai-prompt/localization/custom-messages.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [`NgModule`](link:site.data.urls.angular['ngmodules']) for the Conversational UI components.
 *
 * @example
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { ConversationalUIModule } from '@progress/kendo-angular-conversational-ui';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *   declarations: [AppComponent],
 *   imports: [BrowserModule, ConversationalUIModule],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class ConversationalUIModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ConversationalUIModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: ConversationalUIModule, imports: [i1.AIPromptComponent, i2.PromptViewComponent, i3.OutputViewComponent, i4.CommandViewComponent, i5.CustomViewComponent, i6.AIPromptCustomMessagesComponent, i7.AIPromptToolbarActionsDirective, i8.AIPromptToolbarFocusableDirective, i9.AIPromptOutputTemplateDirective, i10.AIPromptOutputBodyTemplateDirective, i11.ChatComponent, i12.CustomMessagesComponent, i13.AttachmentTemplateDirective, i14.AuthorMessageContentTemplateDirective, i15.ReceiverMessageContentTemplateDirective, i16.MessageContentTemplateDirective, i17.AuthorMessageTemplateDirective, i18.ReceiverMessageTemplateDirective, i19.MessageTemplateDirective, i20.HeroCardComponent, i21.ChatMessageBoxTemplateDirective, i22.ChatHeaderTemplateDirective, i23.NoDataTemplateDirective, i24.ChatTimestampTemplateDirective, i25.ChatStatusTemplateDirective, i26.ChatSuggestionTemplateDirective, i27.ChatUserStatusTemplateDirective, i28.InlineAIPromptComponent, i29.InlineAIPromptOutputTemplateDirective, i30.InlineAIPromptCustomMessagesComponent], exports: [i1.AIPromptComponent, i2.PromptViewComponent, i3.OutputViewComponent, i4.CommandViewComponent, i5.CustomViewComponent, i6.AIPromptCustomMessagesComponent, i7.AIPromptToolbarActionsDirective, i8.AIPromptToolbarFocusableDirective, i9.AIPromptOutputTemplateDirective, i10.AIPromptOutputBodyTemplateDirective, i11.ChatComponent, i12.CustomMessagesComponent, i13.AttachmentTemplateDirective, i14.AuthorMessageContentTemplateDirective, i15.ReceiverMessageContentTemplateDirective, i16.MessageContentTemplateDirective, i17.AuthorMessageTemplateDirective, i18.ReceiverMessageTemplateDirective, i19.MessageTemplateDirective, i20.HeroCardComponent, i21.ChatMessageBoxTemplateDirective, i22.ChatHeaderTemplateDirective, i23.NoDataTemplateDirective, i24.ChatTimestampTemplateDirective, i25.ChatStatusTemplateDirective, i26.ChatSuggestionTemplateDirective, i27.ChatUserStatusTemplateDirective, i28.InlineAIPromptComponent, i29.InlineAIPromptOutputTemplateDirective, i30.InlineAIPromptCustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ConversationalUIModule, providers: [IconsService, PopupService, ResizeBatchService, DialogContainerService, DialogService, WindowService, WindowContainerService], imports: [i1.AIPromptComponent, i2.PromptViewComponent, i3.OutputViewComponent, i4.CommandViewComponent, i11.ChatComponent, i20.HeroCardComponent, i28.InlineAIPromptComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ConversationalUIModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_CONVERSATIONALUI],
                    imports: [...KENDO_CONVERSATIONALUI],
                    providers: [IconsService, PopupService, ResizeBatchService, DialogContainerService, DialogService, WindowService, WindowContainerService]
                }]
        }] });
