// @progress/kendo-angular-conversational-ui/esm2022/ai-prompt/aiprompt.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from "@progress/kendo-angular-popup";
import { ResizeBatchService } from "@progress/kendo-angular-common";
import { DialogContainerService, DialogService, WindowContainerService, WindowService } from "@progress/kendo-angular-dialog";
import { KENDO_AIPROMPT } from "../directives";
import * as i0 from "@angular/core";
import * as i1 from "./aiprompt.component";
import * as i2 from "./views/prompt-view.component";
import * as i3 from "./views/output-view.component";
import * as i4 from "./views/command-view.component";
import * as i5 from "./views/custom-view.component";
import * as i6 from "./localization/custom-messages.component";
import * as i7 from "./templates/toolbar-actions.template";
import * as i8 from "./common/toolbar-focusable.directive";
import * as i9 from "./templates/aiprompt-output-template.directive";
import * as i10 from "./templates/aiprompt-output-body-template.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [`NgModule`](link:site.data.urls.angular['ngmodules']) for the AIPrompt component.
 *
 * @example
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { AIPromptModule } from '@progress/kendo-angular-conversational-ui';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *   imports: [BrowserModule, AIPromptModule],
 *   declarations: [AppComponent],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class AIPromptModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AIPromptModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: AIPromptModule, imports: [i1.AIPromptComponent, i2.PromptViewComponent, i3.OutputViewComponent, i4.CommandViewComponent, i5.CustomViewComponent, i6.AIPromptCustomMessagesComponent, i7.AIPromptToolbarActionsDirective, i8.AIPromptToolbarFocusableDirective, i9.AIPromptOutputTemplateDirective, i10.AIPromptOutputBodyTemplateDirective], exports: [i1.AIPromptComponent, i2.PromptViewComponent, i3.OutputViewComponent, i4.CommandViewComponent, i5.CustomViewComponent, i6.AIPromptCustomMessagesComponent, i7.AIPromptToolbarActionsDirective, i8.AIPromptToolbarFocusableDirective, i9.AIPromptOutputTemplateDirective, i10.AIPromptOutputBodyTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AIPromptModule, providers: [IconsService, PopupService, ResizeBatchService, DialogContainerService, DialogService, WindowService, WindowContainerService], imports: [i1.AIPromptComponent, i2.PromptViewComponent, i3.OutputViewComponent, i4.CommandViewComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AIPromptModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_AIPROMPT],
                    imports: [...KENDO_AIPROMPT],
                    providers: [IconsService, PopupService, ResizeBatchService, DialogContainerService, DialogService, WindowService, WindowContainerService]
                }]
        }] });
