// @progress/kendo-angular-dateinputs/esm2022/dateinput/dateinput.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_DATEINPUT } from '../directives';
import { IconsService } from '@progress/kendo-angular-icons';
import * as i0 from "@angular/core";
import * as i1 from "./localization/dateinput-custom-messages.component";
import * as i2 from "./dateinput.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Required for adding all DateInput features in NgModule-based Angular applications.
 *
 * The package exports:
 * - `DateInputComponent`&mdash;The DateInput component.
 * - `DateInputMessagesComponent`&mdash;The DateInput custom messages component.
 *
 * @example
 * ```ts
 * import { DateInputModule } from '@progress/kendo-angular-dateinputs';
 *
 * @NgModule({
 *   imports: [DateInputModule]
 * })
 * export class AppModule { }
 * ```
 */
export class DateInputModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: DateInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: DateInputModule, imports: [i1.DateInputCustomMessagesComponent, i2.DateInputComponent], exports: [i1.DateInputCustomMessagesComponent, i2.DateInputComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: DateInputModule, providers: [
            IconsService
        ], imports: [i2.DateInputComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: DateInputModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_DATEINPUT],
                    imports: [...KENDO_DATEINPUT],
                    providers: [
                        IconsService
                    ]
                }]
        }] });
