// @progress/kendo-angular-grid/esm2022/rendering/toolbar/tools/ai-assistant/models.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { HttpHeaders } from "@angular/common/http";
import { PreventableEvent } from "@progress/kendo-angular-common";
/**
 * @hidden
 */
export const DEFAULT_AI_REQUEST_OPTIONS = {
    headers: new HttpHeaders({
        'Content-Type': 'application/json'
    }),
    role: 'user',
    method: 'POST',
    responseType: 'json'
};
/**
 * Represents the event data when the AI Assistant request completes successfully.
 */
export class GridAIAssistantResponseSuccessEvent extends PreventableEvent {
    /**
     * The HTTP response from the AI service.
     */
    response;
    constructor(response) {
        super();
        this.response = response;
    }
}
/**
 * Represents the event data when the AI Assistant request completes with an error.
 */
export class GridAIAssistantResponseErrorEvent extends PreventableEvent {
    /**
     * The HTTP error response from the AI service.
     */
    error;
    constructor(error) {
        super();
        this.error = error;
    }
}
