// @progress/kendo-angular-menu/esm2022/menu.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_MENU } from './directives';
import { PopupService } from '@progress/kendo-angular-popup';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./menu.component";
import * as i2 from "./menu-item.component";
import * as i3 from "./templates/item-template.directive";
import * as i4 from "./templates/item-link-template.directive";
import * as i5 from "./templates/item-content-template.directive";
import * as i6 from "./data-binding/hierachy-binding.directive";
import * as i7 from "./data-binding/flat-binding.directive";
import * as i8 from "./rendering/link.directive";
import * as i9 from "./rendering/arrow.component";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Menu component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the Menu module
 * import { MenuModule } from '@progress/kendo-angular-menu';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, BrowserAnimationsModule, MenuModule], // import Menu module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class MenuModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: MenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: MenuModule, imports: [i1.MenuComponent, i2.MenuItemComponent, i3.ItemTemplateDirective, i4.ItemLinkTemplateDirective, i5.ItemContentTemplateDirective, i6.HierarchyBindingDirective, i7.FlatBindingDirective, i8.LinkDirective, i9.ExpandArrowComponent], exports: [i1.MenuComponent, i2.MenuItemComponent, i3.ItemTemplateDirective, i4.ItemLinkTemplateDirective, i5.ItemContentTemplateDirective, i6.HierarchyBindingDirective, i7.FlatBindingDirective, i8.LinkDirective, i9.ExpandArrowComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: MenuModule, providers: [PopupService, IconsService, ResizeBatchService], imports: [i1.MenuComponent, i9.ExpandArrowComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: MenuModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_MENU],
                    imports: [...KENDO_MENU],
                    providers: [PopupService, IconsService, ResizeBatchService]
                }]
        }] });
