// @progress/kendo-angular-menu/esm2022/menus.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_MENUS } from './directives';
import { PopupService } from '@progress/kendo-angular-popup';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./menu.component";
import * as i2 from "./menu-item.component";
import * as i3 from "./templates/item-template.directive";
import * as i4 from "./templates/item-link-template.directive";
import * as i5 from "./templates/item-content-template.directive";
import * as i6 from "./data-binding/hierachy-binding.directive";
import * as i7 from "./data-binding/flat-binding.directive";
import * as i8 from "./rendering/link.directive";
import * as i9 from "./rendering/arrow.component";
import * as i10 from "./context-menu/context-menu.component";
import * as i11 from "./context-menu/context-menu-template.directive";
import * as i12 from "./context-menu/context-menu-target.directive";
import * as i13 from "./context-menu/context-menu-target-container.directive";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * A [module](link:site.data.urls.angular['ngmoduleapi']) that includes the Menu and ContextMenu components and directives.
 * Imports the MenusModule into your application [root module](link:site.data.urls.angular['ngmodules']#angular-modularity)
 * or any other sub-module that will use the Menu and ContextMenu components.
 *
 * @example
 * ```ts-no-run
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
 * import { MenusModule } from '@progress/kendo-angular-menu';
 * import { AppComponent } from './app.component';
 *
 * _@NgModule({
 *     bootstrap:    [AppComponent],
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, BrowserAnimationsModule, MenusModule]
 * })
 * export class AppModule {
 * }
 * ```
 */
export class MenusModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: MenusModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: MenusModule, imports: [i1.MenuComponent, i2.MenuItemComponent, i3.ItemTemplateDirective, i4.ItemLinkTemplateDirective, i5.ItemContentTemplateDirective, i6.HierarchyBindingDirective, i7.FlatBindingDirective, i8.LinkDirective, i9.ExpandArrowComponent, i10.ContextMenuComponent, i11.ContextMenuTemplateDirective, i12.ContextMenuTargetDirective, i13.ContextMenuTargetContainerDirective, i1.MenuComponent, i2.MenuItemComponent, i3.ItemTemplateDirective, i4.ItemLinkTemplateDirective, i5.ItemContentTemplateDirective, i6.HierarchyBindingDirective, i7.FlatBindingDirective, i8.LinkDirective, i9.ExpandArrowComponent], exports: [i1.MenuComponent, i2.MenuItemComponent, i3.ItemTemplateDirective, i4.ItemLinkTemplateDirective, i5.ItemContentTemplateDirective, i6.HierarchyBindingDirective, i7.FlatBindingDirective, i8.LinkDirective, i9.ExpandArrowComponent, i10.ContextMenuComponent, i11.ContextMenuTemplateDirective, i12.ContextMenuTargetDirective, i13.ContextMenuTargetContainerDirective, i1.MenuComponent, i2.MenuItemComponent, i3.ItemTemplateDirective, i4.ItemLinkTemplateDirective, i5.ItemContentTemplateDirective, i6.HierarchyBindingDirective, i7.FlatBindingDirective, i8.LinkDirective, i9.ExpandArrowComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: MenusModule, providers: [PopupService, IconsService, ResizeBatchService], imports: [i1.MenuComponent, i9.ExpandArrowComponent, i10.ContextMenuComponent, i1.MenuComponent, i9.ExpandArrowComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: MenusModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_MENUS],
                    imports: [...KENDO_MENUS],
                    providers: [PopupService, IconsService, ResizeBatchService]
                }]
        }] });
