// @progress/kendo-angular-conversational-ui/esm2022/inline-ai-prompt/inlineaiprompt.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from "@progress/kendo-angular-popup";
import { ResizeBatchService } from "@progress/kendo-angular-common";
import { DialogContainerService, DialogService, WindowContainerService, WindowService } from "@progress/kendo-angular-dialog";
import { KENDO_INLINEAIPROMPT } from "../directives";
import * as i0 from "@angular/core";
import * as i1 from "./inlineaiprompt.component";
import * as i2 from "./output-template.directive";
import * as i3 from "./localization/custom-messages.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [`NgModule`](link:site.data.urls.angular['ngmodules']) for the InlineAIPrompt component.
 *
 * @example
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { InlineAIPromptModule } from '@progress/kendo-angular-conversational-ui';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *   imports: [BrowserModule, InlineAIPromptModule],
 *   declarations: [AppComponent],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class InlineAIPromptModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: InlineAIPromptModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: InlineAIPromptModule, imports: [i1.InlineAIPromptComponent, i2.InlineAIPromptOutputTemplateDirective, i3.InlineAIPromptCustomMessagesComponent], exports: [i1.InlineAIPromptComponent, i2.InlineAIPromptOutputTemplateDirective, i3.InlineAIPromptCustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: InlineAIPromptModule, providers: [IconsService, PopupService, ResizeBatchService, DialogContainerService, DialogService, WindowService, WindowContainerService], imports: [i1.InlineAIPromptComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: InlineAIPromptModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_INLINEAIPROMPT],
                    imports: [...KENDO_INLINEAIPROMPT],
                    providers: [IconsService, PopupService, ResizeBatchService, DialogContainerService, DialogService, WindowService, WindowContainerService]
                }]
        }] });
