// @progress/kendo-angular-conversational-ui/esm2022/inline-ai-prompt/output-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Optional, TemplateRef } from "@angular/core";
import * as i0 from "@angular/core";
/**
 * Defines a template for displaying the output of the Inline AI Prompt.
 *
 * @example
 * ```html
 * <kendo-inlineaiprompt>
 *   <ng-template kendoInlineAIPromptOutputTemplate let-output>
 *     <div class="custom-output">
 *       <h4>AI Response:</h4>
 *       <p>{{ output.output }}</p>
 *     </div>
 *   </ng-template>
 * </kendo-inlineaiprompt>
 * ```
 */
export class InlineAIPromptOutputTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: InlineAIPromptOutputTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: InlineAIPromptOutputTemplateDirective, isStandalone: true, selector: "[kendoInlineAIPromptOutputTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: InlineAIPromptOutputTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoInlineAIPromptOutputTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }] });
