// @progress/kendo-angular-conversational-ui/esm2022/inline-ai-prompt/utils.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { copyIcon, arrowRotateCwIcon, cancelOutlineIcon } from "@progress/kendo-svg-icons";
/**
 * @hidden
 */
export const defaultPopupSettings = {
    popupAlign: { vertical: 'top', horizontal: 'left' },
    anchorAlign: { horizontal: "left", vertical: "bottom" },
    positionMode: 'absolute',
    collision: { horizontal: 'fit', vertical: 'flip' },
    animate: true,
    offset: { top: 0, left: 0 },
    margin: { horizontal: 0, vertical: 0 },
};
/**
 * @hidden
 */
export const calculateMeasurement = (value) => {
    return value + (typeof value === 'number' ? 'px' : '');
};
/**
 * @hidden
 */
export const defaultOutputActions = [{
        name: 'copy',
        type: 'button',
        icon: 'copy',
        svgIcon: copyIcon,
        text: 'Copy',
        fillMode: 'flat',
        themeColor: 'primary',
        rounded: 'medium',
    },
    {
        name: 'retry',
        type: 'button',
        icon: 'arrow-rotate-cw',
        svgIcon: arrowRotateCwIcon,
        text: 'Retry',
        fillMode: 'flat',
        themeColor: 'primary',
        rounded: 'medium',
    },
    {
        name: 'discard',
        type: 'button',
        icon: 'cancel-outline',
        svgIcon: cancelOutlineIcon,
        text: 'Discard',
        fillMode: 'flat',
        themeColor: 'base',
        rounded: 'medium',
    }
];
