// @progress/kendo-angular-grid/esm2022/column-menu/column-menu-item-base.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Input, isDevMode, HostBinding, Component } from '@angular/core';
import { ColumnMenuService } from './column-menu.service';
import { ColumnMenuErrorMessages } from '../common/error-messages';
import * as i0 from "@angular/core";
/**
 * @hidden
 */
export class ColumnMenuItemBase {
    /**
     * Represents the [`ColumnMenuService`]({% slug api_grid_columnmenuservice %}) class.
     * This input is required.
     */
    service;
    hostClass = true;
    ngOnInit() {
        if (isDevMode() && !this.service) {
            throw new Error(ColumnMenuErrorMessages.serviceInput);
        }
    }
    /**
     * @hidden
     */
    close() {
        this.service.close();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ColumnMenuItemBase, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.14", type: ColumnMenuItemBase, selector: "kendo-grid-column-menu-item-base", inputs: { service: "service" }, host: { properties: { "class.k-columnmenu-item-wrapper": "this.hostClass" } }, ngImport: i0, template: ``, isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ColumnMenuItemBase, decorators: [{
            type: Component,
            args: [{
                    selector: 'kendo-grid-column-menu-item-base',
                    template: ``
                }]
        }], propDecorators: { service: [{
                type: Input
            }], hostClass: [{
                type: HostBinding,
                args: ['class.k-columnmenu-item-wrapper']
            }] } });
