// @progress/kendo-angular-grid/esm2022/column-menu/column-list.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, HostBinding, Input, ElementRef, NgZone, Renderer2, Output, EventEmitter, ViewChildren, QueryList, Optional, ChangeDetectorRef } from '@angular/core';
import { ColumnMenuService } from './column-menu.service';
import { ColumnListKeyboardNavigation } from './column-list-kb-nav.service';
import { Keys, normalizeNumpadKeys } from '@progress/kendo-angular-common';
import { Subscription } from 'rxjs';
import { NgFor, NgIf } from '@angular/common';
import { CheckBoxComponent } from '@progress/kendo-angular-inputs';
import { take } from 'rxjs/operators';
import { ContextService } from '../common/provider.service';
import { AdaptiveGridService } from '../common/adaptiveness.service';
import { ColumnInfoService } from '../common/column-info.service';
import * as i0 from "@angular/core";
import * as i1 from "./column-list-kb-nav.service";
import * as i2 from "../common/column-info.service";
import * as i3 from "../common/adaptiveness.service";
import * as i4 from "../common/provider.service";
/**
 * @hidden
 */
export class ColumnListComponent {
    element;
    ngZone;
    renderer;
    listNavigationService;
    cdr;
    columnInfoService;
    adaptiveGridService;
    ctx;
    className = true;
    get listSizeMd() {
        return !(this.ctx?.grid?.isActionSheetExpanded);
    }
    get listSizeLg() {
        return this.ctx?.grid?.isActionSheetExpanded;
    }
    reset = new EventEmitter();
    apply = new EventEmitter();
    columnChange = new EventEmitter();
    set columns(value) {
        this.listNavigationService.items = this.checkboxes?.toArray();
        this._columns = value.filter(column => column.includeInChooser !== false);
    }
    get columns() {
        return this._columns;
    }
    get checkedCheckboxesLength() {
        return this.columns?.filter(column => column.currentlyChecked).length;
    }
    filteredColumns = [];
    autoSync = true;
    showSelectAll = false;
    isFiltered = false;
    ariaLabel;
    allowHideAll = false;
    applyText;
    resetText;
    isLast;
    isExpanded;
    service;
    filterable = false;
    checkboxes;
    get columnCheckboxes() {
        if (this.showSelectAll) {
            return this.checkboxes?.toArray().slice(1);
        }
        return this.checkboxes?.toArray();
    }
    get lockedColumnCheckboxes() {
        return this.columnCheckboxes?.filter((checkbox, index) => this.filteredColumns[index]?.locked);
    }
    get unlockedColumnCheckboxes() {
        return this.columnCheckboxes?.filter((checkbox, index) => !this.filteredColumns[index]?.locked);
    }
    get checkedCheckboxes() {
        if (this.showSelectAll && !this.isFiltered) {
            return this.checkboxes?.toArray().slice(1).filter(checkbox => checkbox.checkedState).length;
        }
        return this.checkboxes?.filter(checkbox => checkbox.checkedState).length;
    }
    resetButton;
    applyButton;
    filterInput;
    _columns;
    domSubscriptions = new Subscription();
    lastDisabledCheckbox;
    constructor(element, ngZone, renderer, listNavigationService, cdr, columnInfoService, adaptiveGridService, ctx) {
        this.element = element;
        this.ngZone = ngZone;
        this.renderer = renderer;
        this.listNavigationService = listNavigationService;
        this.cdr = cdr;
        this.columnInfoService = columnInfoService;
        this.adaptiveGridService = adaptiveGridService;
        this.ctx = ctx;
    }
    ngOnInit() {
        if (!this.element) {
            return;
        }
        this.ngZone.runOutsideAngular(() => {
            this.domSubscriptions.add(this.renderer.listen(this.element.nativeElement, 'click', (e) => {
                if (e.target.classList.contains('k-checkbox')) {
                    const label = e.target.closest('.k-column-list-item');
                    this.setTabindex(label);
                }
            }));
            this.domSubscriptions.add(this.renderer.listen(this.element.nativeElement, 'keydown', this.onKeydown));
        });
    }
    ngAfterViewInit() {
        this.ngZone.onStable.pipe(take(1)).subscribe(() => {
            this.listNavigationService.items = this.checkboxes.toArray();
            if (this.adaptiveGridService?.viewType !== 'columnMenu') {
                this.listNavigationService.toggle(0, true);
            }
            this.updateDisabled();
        });
        if (this.ctx?.grid?.isActionSheetExpanded) {
            this.domSubscriptions.add(this.adaptiveGridService.animationEnd.subscribe(() => {
                this.listNavigationService.toggle(0, true);
            }));
        }
    }
    ngOnChanges(changes) {
        if (!this.service) {
            return;
        }
        if (changes['isLast'] && this.isLast) {
            this.service.menuTabbingService.lastFocusable = this.applyButton?.nativeElement;
        }
        if (changes['isExpanded'] && this.isExpanded && this.isLast && this.applyButton) {
            this.service.menuTabbingService.lastFocusable = this.applyButton.nativeElement;
        }
    }
    onSelectAllCheckboxChange(checkedState) {
        if (!checkedState && !this.allowHideAll) {
            const columnsToKeep = [];
            if (!columnsToKeep.length) {
                const unlockedColumns = this.filteredColumns.filter(c => !c.locked);
                if (unlockedColumns.length > 0) {
                    columnsToKeep.push(unlockedColumns[0]);
                }
            }
            if (this.lockedColumnCheckboxes.length > 0) {
                const lockedColumns = this.filteredColumns.filter(c => c.locked);
                if (lockedColumns.length > 0) {
                    columnsToKeep.unshift(lockedColumns[0]);
                }
            }
            this.filteredColumns.forEach(column => {
                const isChecked = columnsToKeep.indexOf(column) >= 0;
                column.currentlyChecked = isChecked;
            });
        }
        else {
            this.filteredColumns.forEach(column => {
                column.currentlyChecked = checkedState;
            });
        }
        if (this.autoSync) {
            const changedColumns = this.filteredColumns.filter(column => column.initiallyChecked !== column.currentlyChecked);
            if (changedColumns.length > 0) {
                changedColumns.forEach(column => {
                    column.hidden = !column.currentlyChecked;
                    column.initiallyChecked = column.currentlyChecked;
                });
                this.ngZone.run(() => {
                    this.columnChange.emit(changedColumns);
                });
            }
        }
        this.filteredColumns = [...this.filteredColumns];
        this.updateDisabled();
    }
    ngOnDestroy() {
        this.domSubscriptions.unsubscribe();
    }
    cancelChanges() {
        this.columns.forEach((column) => {
            column.currentlyChecked = !column.hidden;
            column.initiallyChecked = !column.hidden;
        });
        this.filteredColumns.forEach((column) => {
            column.currentlyChecked = !column.hidden;
            column.initiallyChecked = !column.hidden;
        });
        this.updateDisabled();
        this.reset.emit();
    }
    applyChanges() {
        const changed = [];
        this.columns.forEach(column => {
            if (column.initiallyChecked !== column.currentlyChecked) {
                column.hidden = !column.currentlyChecked;
                changed.push(column);
            }
        });
        this.updateDisabled();
        this.apply.emit(changed);
        if (changed.length) {
            this.cdr.markForCheck();
            this.columnInfoService?.changeVisibility(changed);
        }
    }
    focusActiveColumn(e) {
        const keyboardEvent = e;
        if (keyboardEvent.shiftKey) {
            this.listNavigationService.toggle(this.listNavigationService.activeIndex, true);
            e.preventDefault();
        }
    }
    onTab(e) {
        if (this.isLast) {
            e.preventDefault();
            if (this.service) {
                this.service.menuTabbingService.firstFocusable.focus();
            }
            else if (this.filterable) {
                this.filterInput?.input.nativeElement.focus();
            }
            else {
                this.listNavigationService.toggle(this.listNavigationService.activeIndex, true);
            }
        }
    }
    onKeydown = (e) => {
        const code = normalizeNumpadKeys(e);
        if (code !== Keys.Tab) {
            e.preventDefault();
        }
        if (code === 'Tab' && e.shiftKey) {
            this.resetButton?.nativeElement.focus();
            e.preventDefault();
        }
        if (code === 'Tab' && !e.shiftKey && this.autoSync) {
            if (this.filterable) {
                this.filterInput?.input.nativeElement.focus();
            }
            e.preventDefault();
        }
        if (code === 'Tab' && e.shiftKey && this.autoSync && this.filterable) {
            this.filterInput?.input.nativeElement.focus();
            e.preventDefault();
        }
        if (code === 'Tab' && !e.shiftKey && !this.autoSync) {
            this.applyButton?.nativeElement.focus();
            e.preventDefault();
        }
        if (code === 'Tab' && e.shiftKey) {
            if (!this.autoSync && this.filterable && e.target.matches('.k-column-list-item')) {
                this.filterInput?.input.nativeElement.focus();
                e.preventDefault();
                return;
            }
            this.ngZone.run(() => {
                if (e.target.matches('.k-column-list-item')) {
                    e.preventDefault();
                    this.resetButton?.nativeElement.focus();
                }
            });
        }
        if (code === Keys.ArrowDown) {
            this.listNavigationService.next();
        }
        else if (code === Keys.ArrowUp) {
            this.listNavigationService.prev();
        }
        else if (code === Keys.Space && e.target.classList.contains('k-column-list-item')) {
            this.listNavigationService.toggleCheckedState();
        }
    };
    updateDisabled() {
        const hasLockedColumns = this.columns.filter(c => c.locked).length > 0;
        if (this.allowHideAll && !hasLockedColumns) {
            return;
        }
        const currentlyCheckedUnlockedColumns = this.columns.filter(column => !column.locked && column.currentlyChecked).length;
        const shouldEnforceMinimumColumns = hasLockedColumns || !this.allowHideAll;
        if (shouldEnforceMinimumColumns && currentlyCheckedUnlockedColumns === 1) {
            this.disableFirstUnlockedCheckedCheckbox();
        }
        else {
            this.enableLastDisabledColumn();
        }
    }
    disableFirstUnlockedCheckedCheckbox() {
        const index = this.filteredColumns.filter(c => !c.locked).findIndex(column => column.currentlyChecked);
        const firstUnlockedCheckedCheckbox = this.unlockedColumnCheckboxes[index];
        this.setDisabledState(firstUnlockedCheckedCheckbox, true);
        this.lastDisabledCheckbox = firstUnlockedCheckedCheckbox;
    }
    enableLastDisabledColumn() {
        this.setDisabledState(this.lastDisabledCheckbox, false);
        this.lastDisabledCheckbox = null;
    }
    setDisabledState(checkbox, disabled) {
        if (!checkbox) {
            return;
        }
        if (checkbox.disabled !== disabled) {
            this.ngZone.runOutsideAngular(() => {
                checkbox.disabled = disabled;
                const checkboxElement = checkbox.hostElement.nativeElement;
                const parentElement = checkboxElement.parentElement;
                if (disabled) {
                    this.renderer.addClass(parentElement, 'k-disabled');
                    this.renderer.setAttribute(parentElement, 'aria-disabled', 'true');
                }
                else {
                    this.renderer.removeClass(parentElement, 'k-disabled');
                    this.renderer.removeAttribute(parentElement, 'aria-disabled');
                }
            });
        }
    }
    onCheckboxChange(checkedState, column, _index) {
        column.currentlyChecked = checkedState;
        if (this.autoSync) {
            column.hidden = !checkedState;
            column.initiallyChecked = checkedState;
            this.columnChange.emit([column]);
        }
        this.updateDisabled();
    }
    setTabindex(labelElement) {
        if (!labelElement) {
            return;
        }
        const allLabels = this.checkboxes.map(checkbox => checkbox.hostElement.nativeElement.parentElement);
        allLabels.forEach((label) => {
            const input = label.querySelector('input');
            input.classList.remove('k-focus');
            this.renderer.removeAttribute(label, 'tabindex');
        });
        this.renderer.setAttribute(labelElement, 'tabindex', '0');
        labelElement.focus();
        this.listNavigationService.activeIndex = allLabels.indexOf(labelElement);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ColumnListComponent, deps: [{ token: i0.ElementRef }, { token: i0.NgZone }, { token: i0.Renderer2 }, { token: i1.ColumnListKeyboardNavigation }, { token: i0.ChangeDetectorRef }, { token: i2.ColumnInfoService, optional: true }, { token: i3.AdaptiveGridService, optional: true }, { token: i4.ContextService, optional: true }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.14", type: ColumnListComponent, isStandalone: true, selector: "kendo-grid-columnlist", inputs: { columns: "columns", filteredColumns: "filteredColumns", autoSync: "autoSync", showSelectAll: "showSelectAll", isFiltered: "isFiltered", ariaLabel: "ariaLabel", allowHideAll: "allowHideAll", applyText: "applyText", resetText: "resetText", isLast: "isLast", isExpanded: "isExpanded", service: "service", filterable: "filterable" }, outputs: { reset: "reset", apply: "apply", columnChange: "columnChange" }, host: { properties: { "class.k-column-list-wrapper": "this.className", "class.k-column-list-md": "this.listSizeMd", "class.k-column-list-lg": "this.listSizeLg" } }, providers: [ColumnListKeyboardNavigation], viewQueries: [{ propertyName: "checkboxes", predicate: CheckBoxComponent, descendants: true }], usesOnChanges: true, ngImport: i0, template: `
        <div 
            class="k-column-list"
            role="listbox"
            aria-multiselectable="true"
            [attr.aria-label]="ariaLabel">
            <label
                *ngIf="showSelectAll"
                class='k-column-list-item'
                role="option">
                <kendo-checkbox
                    #checkbox
                    [inputAttributes]="{'data-index': '0'}"
                    [tabindex]="-1"
                    [checkedState]="checkedCheckboxesLength === columns.length"
                    (checkedStateChange)="onSelectAllCheckboxChange($event)"
                ></kendo-checkbox>
                <span class="k-checkbox-label">{{ 'Select all' }}</span>
            </label>
            <label
                *ngFor="let column of filteredColumns; let index = index;"
                class='k-column-list-item'
                role="option">
                <kendo-checkbox
                    #checkbox
                    [inputAttributes]="{'data-index': index.toString()}"
                    [tabindex]="-1"
                    [checkedState]="column.currentlyChecked"
                    (checkedStateChange)="onCheckboxChange($event, column, index)"
                ></kendo-checkbox>
                <span class="k-checkbox-label">{{ column.displayTitle }}</span>
            </label>
        </div>
    `, isInline: true, dependencies: [{ kind: "directive", type: NgFor, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: CheckBoxComponent, selector: "kendo-checkbox", inputs: ["checkedState", "rounded"], outputs: ["checkedStateChange"], exportAs: ["kendoCheckBox"] }] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ColumnListComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'kendo-grid-columnlist',
                    providers: [ColumnListKeyboardNavigation],
                    template: `
        <div 
            class="k-column-list"
            role="listbox"
            aria-multiselectable="true"
            [attr.aria-label]="ariaLabel">
            <label
                *ngIf="showSelectAll"
                class='k-column-list-item'
                role="option">
                <kendo-checkbox
                    #checkbox
                    [inputAttributes]="{'data-index': '0'}"
                    [tabindex]="-1"
                    [checkedState]="checkedCheckboxesLength === columns.length"
                    (checkedStateChange)="onSelectAllCheckboxChange($event)"
                ></kendo-checkbox>
                <span class="k-checkbox-label">{{ 'Select all' }}</span>
            </label>
            <label
                *ngFor="let column of filteredColumns; let index = index;"
                class='k-column-list-item'
                role="option">
                <kendo-checkbox
                    #checkbox
                    [inputAttributes]="{'data-index': index.toString()}"
                    [tabindex]="-1"
                    [checkedState]="column.currentlyChecked"
                    (checkedStateChange)="onCheckboxChange($event, column, index)"
                ></kendo-checkbox>
                <span class="k-checkbox-label">{{ column.displayTitle }}</span>
            </label>
        </div>
    `,
                    standalone: true,
                    imports: [NgFor, NgIf, CheckBoxComponent]
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i0.NgZone }, { type: i0.Renderer2 }, { type: i1.ColumnListKeyboardNavigation }, { type: i0.ChangeDetectorRef }, { type: i2.ColumnInfoService, decorators: [{
                    type: Optional
                }] }, { type: i3.AdaptiveGridService, decorators: [{
                    type: Optional
                }] }, { type: i4.ContextService, decorators: [{
                    type: Optional
                }] }], propDecorators: { className: [{
                type: HostBinding,
                args: ["class.k-column-list-wrapper"]
            }], listSizeMd: [{
                type: HostBinding,
                args: ['class.k-column-list-md']
            }], listSizeLg: [{
                type: HostBinding,
                args: ['class.k-column-list-lg']
            }], reset: [{
                type: Output
            }], apply: [{
                type: Output
            }], columnChange: [{
                type: Output
            }], columns: [{
                type: Input
            }], filteredColumns: [{
                type: Input
            }], autoSync: [{
                type: Input
            }], showSelectAll: [{
                type: Input
            }], isFiltered: [{
                type: Input
            }], ariaLabel: [{
                type: Input
            }], allowHideAll: [{
                type: Input
            }], applyText: [{
                type: Input
            }], resetText: [{
                type: Input
            }], isLast: [{
                type: Input
            }], isExpanded: [{
                type: Input
            }], service: [{
                type: Input
            }], filterable: [{
                type: Input
            }], checkboxes: [{
                type: ViewChildren,
                args: [CheckBoxComponent]
            }] } });
