// @progress/kendo-angular-layout/esm2022/avatar.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_AVATAR } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./avatar/avatar.component";
import * as i2 from "./avatar/l10n/custom-messages.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the Avatar component.
 *
 * Use this module to add the Avatar component to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { AvatarModule } from '@progress/kendo-angular-layout';
 * import { NgModule } from '@angular/core';
 *
 * @NgModule({
 *   imports: [AvatarModule]
 * })
 * export class AppModule { }
 * ```
 */
export class AvatarModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AvatarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: AvatarModule, imports: [i1.AvatarComponent, i2.AvatarCustomMessagesComponent], exports: [i1.AvatarComponent, i2.AvatarCustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AvatarModule, imports: [i1.AvatarComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AvatarModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_AVATAR],
                    imports: [...KENDO_AVATAR]
                }]
        }] });
