// @progress/kendo-angular-layout/esm2022/expansionpanel.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_EXPANSIONPANEL } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./expansionpanel/expansionpanel.component";
import * as i2 from "./expansionpanel/expansionpanel-title.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the ExpansionPanel component.
 *
 * Use this module to add the ExpansionPanel component to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { ExpansionPanelModule } from '@progress/kendo-angular-layout';
 * import { NgModule } from '@angular/core';
 *
 * @NgModule({
 *   imports: [ExpansionPanelModule]
 * })
 * export class AppModule { }
 * ```
 */
export class ExpansionPanelModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ExpansionPanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: ExpansionPanelModule, imports: [i1.ExpansionPanelComponent, i2.ExpansionPanelTitleDirective], exports: [i1.ExpansionPanelComponent, i2.ExpansionPanelTitleDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ExpansionPanelModule, imports: [i1.ExpansionPanelComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ExpansionPanelModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_EXPANSIONPANEL],
                    imports: [...KENDO_EXPANSIONPANEL]
                }]
        }] });
