// @progress/kendo-angular-layout/esm2022/gridlayout.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_GRIDLAYOUT } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./layouts/grid-layout.component";
import * as i2 from "./layouts/gridlayout-item.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the GridLayout component.
 *
 * Use this module to add the GridLayout component to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { GridLayoutModule } from '@progress/kendo-angular-layout';
 * import { NgModule } from '@angular/core';
 *
 * @NgModule({
 *   imports: [GridLayoutModule]
 * })
 * export class AppModule { }
 * ```
 */
export class GridLayoutModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: GridLayoutModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: GridLayoutModule, imports: [i1.GridLayoutComponent, i2.GridLayoutItemComponent], exports: [i1.GridLayoutComponent, i2.GridLayoutItemComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: GridLayoutModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: GridLayoutModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_GRIDLAYOUT],
                    imports: [...KENDO_GRIDLAYOUT]
                }]
        }] });
