// @progress/kendo-angular-layout/esm2022/splitter.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_SPLITTER } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./splitter/splitter.component";
import * as i2 from "./splitter/splitter-pane.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the Splitter component.
 *
 * Use this module to add the Splitter component to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { SplitterModule } from '@progress/kendo-angular-layout';
 * import { NgModule } from '@angular/core';
 *
 * @NgModule({
 *   imports: [SplitterModule]
 * })
 * export class AppModule { }
 * ```
 */
export class SplitterModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: SplitterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: SplitterModule, imports: [i1.SplitterComponent, i2.SplitterPaneComponent], exports: [i1.SplitterComponent, i2.SplitterPaneComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: SplitterModule, imports: [i1.SplitterComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: SplitterModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_SPLITTER],
                    imports: [...KENDO_SPLITTER]
                }]
        }] });
