// @progress/kendo-angular-layout/esm2022/tabstrip.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_TABSTRIP } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./tabstrip/tabstrip.component";
import * as i2 from "./tabstrip/models/tabstrip-tab.component";
import * as i3 from "./tabstrip/directives/tab-content.directive";
import * as i4 from "./tabstrip/directives/tab-title.directive";
import * as i5 from "./tabstrip/localization/custom-messages.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the TabStrip component.
 *
 * Use this module to add the TabStrip component to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { TabStripModule } from '@progress/kendo-angular-layout';
 * import { NgModule } from '@angular/core';
 *
 * @NgModule({
 *   imports: [TabStripModule]
 * })
 * export class AppModule { }
 * ```
 */
export class TabStripModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TabStripModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: TabStripModule, imports: [i1.TabStripComponent, i2.TabStripTabComponent, i3.TabContentDirective, i4.TabTitleDirective, i5.TabStripCustomMessagesComponent], exports: [i1.TabStripComponent, i2.TabStripTabComponent, i3.TabContentDirective, i4.TabTitleDirective, i5.TabStripCustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TabStripModule, imports: [i1.TabStripComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TabStripModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_TABSTRIP],
                    imports: [...KENDO_TABSTRIP]
                }]
        }] });
