// @progress/kendo-angular-layout/esm2022/tilelayout.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_TILELAYOUT } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./tilelayout/tilelayout.component";
import * as i2 from "./tilelayout/tilelayout-item-body.component";
import * as i3 from "./tilelayout/tilelayout-item.component";
import * as i4 from "./tilelayout/tilelayout-item-header.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the TileLayout component.
 *
 * Use this module to add the TileLayout component to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { TileLayoutModule } from '@progress/kendo-angular-layout';
 * import { NgModule } from '@angular/core';
 *
 * @NgModule({
 *   imports: [TileLayoutModule]
 * })
 * export class AppModule { }
 * ```
 */
export class TileLayoutModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TileLayoutModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: TileLayoutModule, imports: [i1.TileLayoutComponent, i2.TileLayoutItemBodyComponent, i3.TileLayoutItemComponent, i4.TileLayoutItemHeaderComponent], exports: [i1.TileLayoutComponent, i2.TileLayoutItemBodyComponent, i3.TileLayoutItemComponent, i4.TileLayoutItemHeaderComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TileLayoutModule, imports: [i1.TileLayoutComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TileLayoutModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_TILELAYOUT],
                    imports: [...KENDO_TILELAYOUT]
                }]
        }] });
