// @progress/kendo-angular-layout/esm2022/timeline.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { KENDO_TIMELINE } from "./directives";
import * as i0 from "@angular/core";
import * as i1 from "./timeline/timeline.component";
import * as i2 from "./timeline/localization/custom-messages.component";
import * as i3 from "./timeline/templates/timeline-card-actions.directive";
import * as i4 from "./timeline/templates/timeline-card-body.directive";
import * as i5 from "./timeline/templates/timeline-card-header.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the Timeline component.
 *
 * Use this module to add the Timeline component to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { TimelineModule } from '@progress/kendo-angular-layout';
 * import { NgModule } from '@angular/core';
 *
 * @NgModule({
 *   imports: [TimelineModule]
 * })
 * export class AppModule { }
 * ```
 */
export class TimelineModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TimelineModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: TimelineModule, imports: [i1.TimelineComponent, i2.TimelineCustomMessagesComponent, i3.TimelineCardActionsTemplateDirective, i4.TimelineCardBodyTemplateDirective, i5.TimelineCardHeaderTemplateDirective], exports: [i1.TimelineComponent, i2.TimelineCustomMessagesComponent, i3.TimelineCardActionsTemplateDirective, i4.TimelineCardBodyTemplateDirective, i5.TimelineCardHeaderTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TimelineModule, imports: [i1.TimelineComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TimelineModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_TIMELINE],
                    imports: [...KENDO_TIMELINE]
                }]
        }] });
