// @progress/kendo-angular-dialog/esm2022/window/navigation.service.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Injectable, NgZone } from "@angular/core";
import { DragResizeService } from './drag-resize.service';
import { isPresent, OFFSET_STYLES, WINDOW_CLASSES, hasClasses, WINDOW_ELEMENTS_HANDLING_ESC_KEY } from '../common/util';
import { Keys, normalizeNumpadKeys } from "@progress/kendo-angular-common";
import * as i0 from "@angular/core";
import * as i1 from "./drag-resize.service";
/**
 * @hidden
 */
export class NavigationService {
    window;
    ngZone;
    constructor(window, ngZone) {
        this.window = window;
        this.ngZone = ngZone;
    }
    process(ev) {
        const key = normalizeNumpadKeys(ev);
        const target = ev.target;
        switch (key) {
            case Keys.ArrowUp:
            case Keys.ArrowDown:
            case Keys.ArrowLeft:
            case Keys.ArrowRight: {
                if (hasClasses(target, WINDOW_CLASSES)) {
                    ev.preventDefault();
                    this.ngZone.run(() => {
                        this.handleArrow(key, ev);
                    });
                }
                break;
            }
            case Keys.Escape:
                if (hasClasses(target, WINDOW_ELEMENTS_HANDLING_ESC_KEY)) {
                    this.ngZone.run(() => {
                        this.handleEscape();
                    });
                }
                break;
            default:
                break;
        }
    }
    handleArrow(key, ev) {
        const options = this.window.options;
        if (ev.altKey) {
            this.handleStateChange(key, options.state);
            return;
        }
        if ((ev.ctrlKey || ev.metaKey) && options.state === 'default') {
            this.handleResize(key);
        }
        else {
            this.handleDrag(key);
        }
    }
    handleEscape() {
        this.window.closeAction();
    }
    handleDrag(key) {
        const options = this.window.options;
        if (!options.draggable) {
            return;
        }
        const offset = this.window.currentOffsetAndPosition();
        const restoreOptions = this.window.restoreOptions;
        const ev = {};
        let delta = 10;
        if (key === Keys.ArrowLeft || key === Keys.ArrowUp) {
            delta *= -1;
        }
        switch (key) {
            case Keys.ArrowLeft:
            case Keys.ArrowRight: {
                ev.left = offset.x + delta;
                options.left = ev.left;
                break;
            }
            case Keys.ArrowUp:
            case Keys.ArrowDown: {
                ev.top = offset.y + delta;
                options.top = ev.top;
                break;
            }
            default:
                break;
        }
        if (options.state === 'minimized' && isPresent(restoreOptions)) {
            restoreOptions.left = options.left;
            restoreOptions.top = options.top;
        }
        this.window.change.emit(ev);
    }
    handleResize(key) {
        const options = this.window.options;
        if (!options.resizable) {
            return;
        }
        const offset = this.window.currentOffsetAndPosition();
        let newWidth;
        let newHeight;
        const ev = {};
        let delta = 10;
        if (key === Keys.ArrowLeft || key === Keys.ArrowUp) {
            delta *= -1;
        }
        switch (key) {
            case Keys.ArrowLeft:
            case Keys.ArrowRight: {
                newWidth = offset.width + delta;
                if (newWidth !== options.width && newWidth >= options.minWidth) {
                    ev.width = newWidth;
                }
                break;
            }
            case Keys.ArrowUp:
            case Keys.ArrowDown: {
                newHeight = offset.height + delta;
                if (newHeight !== options.height && newHeight >= options.minHeight) {
                    ev.height = newHeight;
                }
                break;
            }
            default:
                break;
        }
        if (isPresent(ev.width) || isPresent(ev.height)) {
            OFFSET_STYLES.forEach((style) => {
                if (isPresent(ev[style])) {
                    this.window.options[style] = ev[style];
                }
            });
            this.window.change.emit(ev);
        }
    }
    handleStateChange(key, state) {
        if ((state === 'minimized' && key === Keys.ArrowUp) ||
            (state === 'maximized' && key === Keys.ArrowDown)) {
            this.window.restoreAction();
            return;
        }
        if (state === 'default') {
            if (key === Keys.ArrowUp) {
                this.window.maximizeAction();
            }
            else if (key === Keys.ArrowDown) {
                this.window.minimizeAction();
            }
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: NavigationService, deps: [{ token: i1.DragResizeService }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: NavigationService });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: NavigationService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.DragResizeService }, { type: i0.NgZone }] });
