// @progress/kendo-angular-conversational-ui/esm2022/ai-prompt/templates/aiprompt-output-body-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Represents the template for the body of the response in the AIPrompt component. The template context is set to the current output object.
 *
 * @example
 * ```html
 * <kendo-aiprompt>
 *   <ng-template kendoAIPromptOutputBodyTemplate let-output>
 *     <div>{{output.prompt}}</div>
 *   </ng-template>
 * </kendo-aiprompt>
 * ```
 */
export class AIPromptOutputBodyTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AIPromptOutputBodyTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: AIPromptOutputBodyTemplateDirective, isStandalone: true, selector: "[kendoAIPromptOutputBodyTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AIPromptOutputBodyTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoAIPromptOutputBodyTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
