// @progress/kendo-angular-conversational-ui/esm2022/ai-prompt/templates/toolbar-actions.template.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Optional, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Represents a template for additional Toolbar actions.
 * The actions appear after the Toolbar buttons for all views.
 *
 * To define the template, nest an `<ng-template>` tag
 * with the `kendoAIPromptToolbarActionsTemplate` directive inside the `<kendo-aiprompt>` tag.
 *
 * @example
 * ```html
 * <kendo-aiprompt>
 *   <ng-template kendoAIPromptToolbarActionsTemplate>
 *     <button kendoButton>Custom Action</button>
 *   </ng-template>
 * </kendo-aiprompt>
 * ```
 */
export class AIPromptToolbarActionsDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AIPromptToolbarActionsDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: AIPromptToolbarActionsDirective, isStandalone: true, selector: "[kendoAIPromptToolbarActionsTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: AIPromptToolbarActionsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoAIPromptToolbarActionsTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }] });
