// @progress/kendo-angular-grid/codemods/utils.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tsInterfaceTransformer = exports.tsComponentPropertyRemoval = exports.templateAttributeRemoval = exports.htmlAttributeRemoval = exports.htmlAttributeValueTransformer = exports.tsPropertyValueTransformer = exports.templateAttributeValueTransformer = exports.htmlEventTransformer = exports.htmlAttributeTransformer = exports.htmlBoundAttributeTransformer = exports.htmlStaticAttributeTransformer = exports.tsPropertyTransformer = exports.templateAttributeTransformer = exports.templateBoundAttributeTransformer = exports.templateStaticAttributeTransformer = exports.templateEventTransformer = exports.blockTextElements = void 0;
exports.hasKendoInTemplate = hasKendoInTemplate;
exports.tsPropertyRemoval = tsPropertyRemoval;
const node_html_parser_1 = __importDefault(require("node-html-parser"));
exports.blockTextElements = {
    script: true,
    noscript: true,
    style: true,
    pre: true,
};
function hasKendoInTemplate(source) {
    const kendoPattern = /(<kendo-[^>\s]+|<[^>]*\s+kendo[A-Z][^>\s]*)/g;
    return kendoPattern.test(source);
}
const templateEventTransformer = (root, tagName, eventName, newEventName) => {
    const elements = Array.from(root.getElementsByTagName(tagName)) || [];
    for (const element of elements) {
        // Handle event bindings like (actionClick)="handler($event)"
        const eventAttr = element.getAttribute(`(${eventName})`);
        if (eventAttr) {
            element.setAttribute(`(${newEventName})`, eventAttr);
            element.removeAttribute(`(${eventName})`);
        }
    }
};
exports.templateEventTransformer = templateEventTransformer;
const templateStaticAttributeTransformer = (root, tagName, attributeName, newAttributeName) => {
    const elements = Array.from(root.getElementsByTagName(tagName)) || [];
    for (const element of elements) {
        // Handle static attributes like title="foo"
        const staticAttr = element.getAttribute(attributeName);
        if (staticAttr) {
            element.setAttribute(newAttributeName, staticAttr);
            element.removeAttribute(attributeName);
        }
    }
};
exports.templateStaticAttributeTransformer = templateStaticAttributeTransformer;
const templateBoundAttributeTransformer = (root, tagName, attributeName, newAttributeName, valueProperty) => {
    const elements = Array.from(root.getElementsByTagName(tagName)) || [];
    for (const element of elements) {
        // Handle bound attributes like [title]="foo" or [title]="'foo'"
        const boundAttr = element.getAttribute(`[${attributeName}]`);
        if (boundAttr) {
            const newValue = valueProperty ? `${boundAttr}.${valueProperty}` : boundAttr;
            element.setAttribute(`[${newAttributeName}]`, newValue);
            element.removeAttribute(`[${attributeName}]`);
        }
    }
};
exports.templateBoundAttributeTransformer = templateBoundAttributeTransformer;
const templateAttributeTransformer = (root, tagName, attributeName, newAttributeName) => {
    (0, exports.templateBoundAttributeTransformer)(root, tagName, attributeName, newAttributeName);
    (0, exports.templateStaticAttributeTransformer)(root, tagName, attributeName, newAttributeName);
};
exports.templateAttributeTransformer = templateAttributeTransformer;
const tsPropertyTransformer = (source, root, j, componentType, propertyName, newPropertyName, valueProperty) => {
    if (source.includes(componentType)) {
        // Find all class properties that are of type DropDownListComponent
        const properties = new Set();
        // Find properties with type annotations
        root.find(j.ClassProperty, {
            typeAnnotation: {
                typeAnnotation: {
                    typeName: {
                        name: componentType,
                    },
                },
            },
        }).forEach((path) => {
            if (path.node.key.type === 'Identifier') {
                properties.add(path.node.key.name);
            }
        });
        // Find function parameters of type componentType
        const parameters = new Set();
        root.find(j.FunctionDeclaration).forEach((path) => {
            if (path.node.params) {
                path.node.params.forEach((param) => {
                    if (param.type === 'Identifier' &&
                        param.typeAnnotation &&
                        param.typeAnnotation.typeAnnotation?.type === 'TSTypeReference' &&
                        param.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                        param.typeAnnotation.typeAnnotation.typeName.name === componentType) {
                        parameters.add(param.name);
                    }
                });
            }
        });
        // Also check method declarations in classes
        root.find(j.ClassMethod).forEach((path) => {
            if (path.node.params) {
                path.node.params.forEach((param) => {
                    if (param.type === 'Identifier' &&
                        param.typeAnnotation &&
                        param.typeAnnotation.typeAnnotation?.type === 'TSTypeReference' &&
                        param.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                        param.typeAnnotation.typeAnnotation.typeName.name === componentType) {
                        parameters.add(param.name);
                    }
                });
            }
        });
        // Also check arrow functions
        root.find(j.ArrowFunctionExpression).forEach((path) => {
            if (path.node.params) {
                path.node.params.forEach((param) => {
                    if (param.type === 'Identifier' &&
                        param.typeAnnotation &&
                        param.typeAnnotation.typeAnnotation?.type === 'TSTypeReference' &&
                        param.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                        param.typeAnnotation.typeAnnotation.typeName.name === componentType) {
                        parameters.add(param.name);
                    }
                });
            }
        });
        // Find local variable declarations of type componentType
        const localVariables = new Set();
        root.find(j.VariableDeclarator).forEach((path) => {
            if (path.node.id.type === 'Identifier' &&
                path.node.id.typeAnnotation &&
                path.node.id.typeAnnotation.typeAnnotation?.type === 'TSTypeReference' &&
                path.node.id.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                path.node.id.typeAnnotation.typeAnnotation.typeName.name === componentType) {
                localVariables.add(path.node.id.name);
            }
        });
        // Find all member expressions where title property is accessed on any componentType instance
        root.find(j.MemberExpression, {
            property: {
                type: 'Identifier',
                name: propertyName,
            },
        })
            .filter((path) => {
            // Filter to only include accesses on properties that are componentType instances
            if (path.node.object.type === 'MemberExpression' && path.node.object.property.type === 'Identifier') {
                // handle properties of this
                if (path.node.object.object.type === 'ThisExpression' &&
                    properties.has(path.node.object.property.name)) {
                    return true;
                }
            }
            // Handle function parameters and local variables
            if (path.node.object.type === 'Identifier') {
                return parameters.has(path.node.object.name) || localVariables.has(path.node.object.name);
            }
            return false;
        })
            .forEach((path) => {
            // Replace old property name with new property name
            if (path.node.property.type === 'Identifier') {
                path.node.property.name = newPropertyName;
            }
            // If valueProperty is specified and this is part of an assignment,
            // we need to also modify the right-hand side of the assignment
            if (valueProperty) {
                const assignmentExpression = path.parent;
                if (assignmentExpression &&
                    assignmentExpression.value &&
                    assignmentExpression.value.type === 'AssignmentExpression' &&
                    assignmentExpression.value.left === path.node) {
                    const rightSide = assignmentExpression.value.right;
                    // Case 1: Right side is a member expression (e.g., this.user, obj.user) -> transform to this.user.id, obj.user.id
                    // Case 2: Right side is an identifier (e.g., user, foo) -> transform to user.id, foo.id
                    if (rightSide.type === 'MemberExpression' || rightSide.type === 'Identifier') {
                        const newRightSide = j.memberExpression(rightSide, j.identifier(valueProperty));
                        assignmentExpression.value.right = newRightSide;
                    }
                    // Case 3: Right side is object literal -> extract the valueProperty value
                    else if (rightSide.type === 'ObjectExpression') {
                        // Find the property that matches valueProperty
                        const targetProperty = rightSide.properties.find((prop) => prop.type === 'ObjectProperty' &&
                            prop.key &&
                            prop.key.type === 'Identifier' &&
                            prop.key.name === valueProperty);
                        if (targetProperty) {
                            // Replace the entire object literal with just the value of the target property
                            assignmentExpression.value.right = targetProperty.value;
                        }
                    }
                }
            }
        });
    }
};
exports.tsPropertyTransformer = tsPropertyTransformer;
const htmlStaticAttributeTransformer = (fileInfo, tagName, oldName, newName) => {
    const fileContent = fileInfo.source;
    const root = (0, node_html_parser_1.default)(fileContent, { comment: true, blockTextElements: exports.blockTextElements });
    let modified = false;
    const elements = Array.from(root.querySelectorAll(tagName));
    for (const element of elements) {
        const staticAttr = element.getAttribute(oldName);
        if (staticAttr) {
            element.removeAttribute(oldName);
            element.setAttribute(newName, staticAttr);
            modified = true;
        }
    }
    if (modified) {
        return root.toString();
    }
    return fileContent;
};
exports.htmlStaticAttributeTransformer = htmlStaticAttributeTransformer;
const htmlBoundAttributeTransformer = (fileInfo, tagName, oldName, newName, valueProperty) => {
    const fileContent = fileInfo.source;
    const root = (0, node_html_parser_1.default)(fileContent, { comment: true, blockTextElements: exports.blockTextElements });
    let modified = false;
    const elements = Array.from(root.querySelectorAll(tagName));
    for (const element of elements) {
        const boundAttr = element.getAttribute(`[${oldName}]`);
        if (boundAttr) {
            element.removeAttribute(`[${oldName}]`);
            // If valueProperty is provided, append it to the bound value
            const newValue = valueProperty ? `${boundAttr}.${valueProperty}` : boundAttr;
            element.setAttribute(`[${newName}]`, newValue);
            modified = true;
        }
    }
    if (modified) {
        return root.toString();
    }
    return fileContent;
};
exports.htmlBoundAttributeTransformer = htmlBoundAttributeTransformer;
const htmlAttributeTransformer = (fileInfo, tagName, oldName, newName) => {
    let content = (0, exports.htmlBoundAttributeTransformer)(fileInfo, tagName, oldName, newName);
    content = (0, exports.htmlStaticAttributeTransformer)({ path: fileInfo.path, source: content }, tagName, oldName, newName);
    return content;
};
exports.htmlAttributeTransformer = htmlAttributeTransformer;
const htmlEventTransformer = (fileInfo, tagName, oldEventName, newEventName) => {
    const fileContent = fileInfo.source;
    const root = (0, node_html_parser_1.default)(fileContent, { comment: true, blockTextElements: exports.blockTextElements });
    let modified = false;
    const elements = Array.from(root.querySelectorAll(tagName));
    for (const element of elements) {
        // Handle event bindings like (actionClick)="handler($event)"
        const eventAttr = element.getAttribute(`(${oldEventName})`);
        if (eventAttr) {
            element.removeAttribute(`(${oldEventName})`);
            element.setAttribute(`(${newEventName})`, eventAttr);
            modified = true;
        }
    }
    if (modified) {
        return root.toString();
    }
    return fileContent;
};
exports.htmlEventTransformer = htmlEventTransformer;
const templateAttributeValueTransformer = (root, tagName, attributeName, oldAttributeValue, newAttributeValue) => {
    const elements = Array.from(root.getElementsByTagName(tagName)) || [];
    for (const element of elements) {
        // Handle bound attributes (e.g., [showText]="'overflow'")
        const boundAttr = element.getAttribute(`[${attributeName}]`);
        if (boundAttr === `'${oldAttributeValue}'`) {
            // For bound literals like [showText]="'overflow'" or [showText]="\"overflow\""
            element.setAttribute(`[${attributeName}]`, boundAttr.replace(oldAttributeValue, newAttributeValue));
        }
        // Handle static attributes like title="foo"
        const staticAttrValue = element.getAttribute(attributeName);
        if (staticAttrValue === oldAttributeValue) {
            element.setAttribute(attributeName, newAttributeValue);
        }
    }
};
exports.templateAttributeValueTransformer = templateAttributeValueTransformer;
const tsPropertyValueTransformer = (source, root, j, typeName, oldValue, newValue) => {
    if (source.includes(typeName)) {
        root.find(j.ClassProperty)
            .filter((path) => {
            if (path.node.typeAnnotation?.typeAnnotation &&
                path.node.typeAnnotation.typeAnnotation.type === 'TSTypeReference' &&
                path.node.typeAnnotation.typeAnnotation.typeName &&
                path.node.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                path.node.typeAnnotation.typeAnnotation.typeName.name === typeName) {
                return true;
            }
            return false;
        })
            .forEach((path) => {
            if (path.node.value && path.node.value.type === 'StringLiteral' && path.node.value.value === oldValue) {
                path.node.value.value = newValue;
            }
        });
        root.find(j.VariableDeclarator)
            .filter((path) => {
            if (path.node.id.type === 'Identifier' &&
                path.node.id.typeAnnotation?.typeAnnotation &&
                path.node.id.typeAnnotation.typeAnnotation.type === 'TSTypeReference' &&
                path.node.id.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                path.node.id.typeAnnotation.typeAnnotation.typeName.name === typeName) {
                return true;
            }
            return false;
        })
            .forEach((path) => {
            if (path.node.id.type === 'Identifier') {
                if (path.node.init && path.node.init.type === 'StringLiteral' && path.node.init.value === oldValue) {
                    path.node.init.value = newValue;
                }
            }
        });
        root.find(j.AssignmentExpression)
            .filter((path) => {
            return path.node.right.type === 'StringLiteral' && path.node.right.value === oldValue;
        })
            .forEach((path) => {
            path.node.right.value = newValue;
        });
        root.find(j.JSXAttribute, {
            value: {
                type: 'StringLiteral',
                value: oldValue,
            },
        }).forEach((path) => {
            if (path.node.value?.type === 'StringLiteral') {
                path.node.value.value = newValue;
            }
        });
    }
};
exports.tsPropertyValueTransformer = tsPropertyValueTransformer;
const htmlAttributeValueTransformer = (fileInfo, tagName, attributeName, oldValue, newValue) => {
    const fileContent = fileInfo.source;
    const root = (0, node_html_parser_1.default)(fileContent, { comment: true, blockTextElements: exports.blockTextElements });
    const elements = root.querySelectorAll(tagName);
    let modified = false;
    for (const element of elements) {
        // Handle static attributes (e.g., showText="overflow")
        const staticAttr = element.getAttribute(attributeName);
        if (staticAttr === oldValue) {
            element.setAttribute(attributeName, newValue);
            modified = true;
            console.log(`Modified static attribute ${attributeName} from "${oldValue}" to "${newValue}" in element:`, element.toString().substring(0, 100));
        }
        // Handle bound attributes (e.g., [showText]="overflow")
        const boundAttr = element.getAttribute(`[${attributeName}]`);
        if (boundAttr) {
            // For bound literals like [showText]="'overflow'" or [showText]="\"overflow\""
            if (boundAttr === `'${oldValue}'` || boundAttr === `"${oldValue}"`) {
                const updatedValue = boundAttr.replace(oldValue, newValue);
                element.setAttribute(`[${attributeName}]`, updatedValue);
                modified = true;
                console.log(`Modified bound attribute [${attributeName}] from "${boundAttr}" to "${updatedValue}" in element:`, element.toString().substring(0, 100));
            }
        }
    }
    if (modified) {
        const updatedContent = root.toString();
        return updatedContent;
    }
    return fileContent;
};
exports.htmlAttributeValueTransformer = htmlAttributeValueTransformer;
const htmlAttributeRemoval = (fileInfo, tagName, attributeName, propertyToRemove) => {
    const filePath = fileInfo.path;
    const fileContent = fileInfo.source;
    const root = (0, node_html_parser_1.default)(fileContent, { comment: true, blockTextElements: exports.blockTextElements });
    // Use the same logic as templateAttributeRemoval
    const elements = root.querySelectorAll(tagName);
    for (const element of elements) {
        // If no propertyToRemove is specified, remove the entire attribute
        if (!propertyToRemove) {
            // Remove both bound and static attributes
            element.removeAttribute(`[${attributeName}]`);
            element.removeAttribute(attributeName);
            continue;
        }
        // Look for bound attribute (e.g., [scrollable]="...")
        const boundAttr = element.getAttribute(`[${attributeName}]`);
        if (boundAttr) {
            if (boundAttr.trim().startsWith('{') && boundAttr.trim().endsWith('}')) {
                const objectLiteral = boundAttr.trim();
                const propRegex = new RegExp(`\\s*${propertyToRemove}\\s*:\\s*[^,}]+\\s*(,\\s*)?`, 'g');
                let newObjectLiteral = objectLiteral.replace(propRegex, '');
                newObjectLiteral = newObjectLiteral.replace(/,\s*}$/, '}');
                if (newObjectLiteral === '{}') {
                    element.removeAttribute(`[${attributeName}]`);
                }
                else {
                    element.setAttribute(`[${attributeName}]`, newObjectLiteral);
                }
            }
            else {
                console.warn(`Cannot remove property from variable reference: ${boundAttr} in file ${filePath}`);
            }
        }
    }
    return root.toString();
};
exports.htmlAttributeRemoval = htmlAttributeRemoval;
const templateAttributeRemoval = (root, tagName, attributeName, propertyToRemove) => {
    const elements = root.querySelectorAll(tagName);
    for (const element of elements) {
        if (!propertyToRemove) {
            element.removeAttribute(`[${attributeName}]`);
            element.removeAttribute(attributeName);
            continue;
        }
        const boundAttr = element.getAttribute(`[${attributeName}]`);
        if (boundAttr) {
            if (boundAttr.trim().startsWith('{') && boundAttr.trim().endsWith('}')) {
                const objectLiteral = boundAttr.trim();
                const propRegex = new RegExp(`\\s*${propertyToRemove}\\s*:\\s*[^,}]+\\s*(,\\s*)?`, 'g');
                let newObjectLiteral = objectLiteral.replace(propRegex, '');
                newObjectLiteral = newObjectLiteral.replace(/,\s*}$/, '}');
                if (newObjectLiteral === '{}') {
                    element.removeAttribute(`[${attributeName}]`);
                }
                else {
                    element.setAttribute(`[${attributeName}]`, newObjectLiteral);
                }
            }
            else {
                console.warn(`Cannot remove property from variable reference: ${boundAttr}`);
            }
        }
    }
};
exports.templateAttributeRemoval = templateAttributeRemoval;
function tsPropertyRemoval(source, rootSource, j, typeName, propertyName) {
    if (source.includes(typeName)) {
        rootSource
            .find(j.ClassProperty, {
            typeAnnotation: {
                typeAnnotation: {
                    typeName: {
                        name: typeName,
                    },
                },
            },
        })
            .forEach((path) => {
            if (path.node.value && path.node.value.type === 'ObjectExpression') {
                const properties = path.node.value.properties;
                const propIndex = properties.findIndex((p) => p.type === 'ObjectProperty' &&
                    p.key &&
                    p.key.type === 'Identifier' &&
                    p.key.name === propertyName);
                if (propIndex !== -1) {
                    if (properties.length === 1) {
                        j(path).remove();
                    }
                    else {
                        properties.splice(propIndex, 1);
                    }
                }
            }
        });
        rootSource
            .find(j.AssignmentExpression, {
            left: {
                type: 'MemberExpression',
                object: {
                    type: 'MemberExpression',
                },
                property: {
                    name: propertyName,
                },
            },
        })
            .forEach((path) => {
            j(path).remove();
        });
        return rootSource;
    }
}
const tsComponentPropertyRemoval = (source, root, j, componentType, componentProperty, propertyToRemove) => {
    if (source.includes(componentType)) {
        // If no propertyToRemove is specified, remove the entire componentProperty
        if (!propertyToRemove) {
            // Handle direct property assignments like: foo.scrollable = value;
            root.find(j.AssignmentExpression)
                .filter((path) => {
                const { left } = path.value;
                // Check if this assigns to component.componentProperty
                if (left &&
                    left.type === 'MemberExpression' &&
                    left.property &&
                    left.property.name === componentProperty) {
                    // Check if the base object is our component type
                    return isComponentTypeMatch(root, j, left.object, componentType);
                }
                return false;
            })
                .forEach((path) => {
                // Remove the entire statement
                j(path).closest(j.ExpressionStatement).remove();
            });
            return root;
        }
        // CASE 1: Handle direct property assignments like: foo.scrollable.mouseScrollSpeed = 3000;
        root.find(j.AssignmentExpression)
            .filter((path) => {
            const { left } = path.value;
            // Check if this is a member expression assignment
            if (left && left.type === 'MemberExpression') {
                // Check if we're accessing the property to remove
                if (left.property && left.property.name === propertyToRemove) {
                    // Check if we're accessing it from component.componentProperty
                    const obj = left.object;
                    if (obj &&
                        obj.type === 'MemberExpression' &&
                        obj.property &&
                        obj.property.name === componentProperty) {
                        // Now check if the base object is our component type
                        return isComponentTypeMatch(root, j, obj.object, componentType);
                    }
                }
            }
            return false;
        })
            .forEach((path) => {
            // Remove the entire statement
            j(path).closest(j.ExpressionStatement).remove();
        });
        // CASE 2 & 3: Handle object assignments like: foo.scrollable = { mouseScrollSpeed: 3000, ... };
        root.find(j.AssignmentExpression)
            .filter((path) => {
            const { left, right } = path.value;
            // Check if this assigns to component.componentProperty
            if (left &&
                left.type === 'MemberExpression' &&
                left.property &&
                left.property.name === componentProperty &&
                right &&
                right.type === 'ObjectExpression') {
                // Check if the base object is our component type
                return isComponentTypeMatch(root, j, left.object, componentType);
            }
            return false;
        })
            .forEach((path) => {
            const properties = path.value.right.properties;
            // Find the property we want to remove
            const propIndex = properties.findIndex((p) => p &&
                p.type === 'ObjectProperty' &&
                p.key &&
                p.key.type === 'Identifier' &&
                p.key.name === propertyToRemove);
            if (propIndex !== -1) {
                // Case 2: If it's the only property, remove the entire statement
                if (properties.length === 1) {
                    j(path).closest(j.ExpressionStatement).remove();
                }
                // Case 3: If there are other properties, just remove this one property
                else {
                    properties.splice(propIndex, 1);
                }
            }
        });
        return root;
    }
};
exports.tsComponentPropertyRemoval = tsComponentPropertyRemoval;
// Helper function to check if a node is a component of the specified type
function isComponentTypeMatch(root, j, node, componentType) {
    if (!node)
        return false;
    if (node.type === 'ThisExpression') {
        return true;
    }
    if (node.type === 'Identifier') {
        const paramName = node.name;
        return root.find(j.Function).some((path) => {
            return (path.node.params &&
                path.node.params.some((param) => param.type === 'Identifier' &&
                    param.name === paramName &&
                    param.typeAnnotation?.typeAnnotation?.typeName?.name === componentType));
        });
    }
    if (node.type === 'MemberExpression') {
        if (node.object.type === 'ThisExpression' && node.property.type === 'Identifier') {
            const propName = node.property.name;
            return (root
                .find(j.ClassProperty, {
                key: { name: propName },
                typeAnnotation: {
                    typeAnnotation: {
                        typeName: { name: componentType },
                    },
                },
            })
                .size() > 0);
        }
    }
    return false;
}
const tsInterfaceTransformer = (fileInfo, rootSource, j, packageName, interfaceName, newName) => {
    const source = fileInfo.source;
    if (source.includes(interfaceName)) {
        let isImportedFromKendoListbox = false;
        rootSource.find(j.ImportDeclaration).forEach((path) => {
            if (path.node.source &&
                path.node.source.value === packageName &&
                path.node.specifiers) {
                path.node.specifiers.forEach((specifier) => {
                    if (specifier.type === 'ImportSpecifier' &&
                        specifier.imported.type === 'Identifier' &&
                        specifier.imported.name === interfaceName) {
                        isImportedFromKendoListbox = true;
                        specifier.imported.name = newName;
                    }
                });
            }
        });
        if (!isImportedFromKendoListbox) {
            return;
        }
        rootSource.find(j.ClassProperty).forEach((path) => {
            if (path.node.typeAnnotation &&
                path.node.typeAnnotation.typeAnnotation &&
                path.node.typeAnnotation.typeAnnotation.type === 'TSTypeReference' &&
                path.node.typeAnnotation.typeAnnotation.typeName &&
                path.node.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                path.node.typeAnnotation.typeAnnotation.typeName.name === interfaceName) {
                path.node.typeAnnotation.typeAnnotation.typeName.name = newName;
            }
        });
        rootSource.find(j.VariableDeclarator).forEach((path) => {
            if (path.node.id.type === 'Identifier' &&
                path.node.id.typeAnnotation &&
                path.node.id.typeAnnotation.typeAnnotation &&
                path.node.id.typeAnnotation.typeAnnotation.type === 'TSTypeReference' &&
                path.node.id.typeAnnotation.typeAnnotation.typeName &&
                path.node.id.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                path.node.id.typeAnnotation.typeAnnotation.typeName.name === interfaceName) {
                path.node.id.typeAnnotation.typeAnnotation.typeName.name = newName;
            }
        });
        rootSource.find(j.FunctionDeclaration).forEach((path) => {
            if (path.node.params) {
                path.node.params.forEach((param) => {
                    if (param.type === 'Identifier' &&
                        param.typeAnnotation &&
                        param.typeAnnotation.typeAnnotation &&
                        param.typeAnnotation.typeAnnotation.type === 'TSTypeReference' &&
                        param.typeAnnotation.typeAnnotation.typeName &&
                        param.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                        param.typeAnnotation.typeAnnotation.typeName.name === interfaceName) {
                        param.typeAnnotation.typeAnnotation.typeName.name = newName;
                    }
                });
            }
        });
        rootSource.find(j.ArrowFunctionExpression).forEach((path) => {
            if (path.node.params) {
                path.node.params.forEach((param) => {
                    if (param.type === 'Identifier' &&
                        param.typeAnnotation &&
                        param.typeAnnotation.typeAnnotation &&
                        param.typeAnnotation.typeAnnotation.type === 'TSTypeReference' &&
                        param.typeAnnotation.typeAnnotation.typeName &&
                        param.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                        param.typeAnnotation.typeAnnotation.typeName.name === interfaceName) {
                        param.typeAnnotation.typeAnnotation.typeName.name = newName;
                    }
                });
            }
        });
        rootSource.find(j.ClassMethod).forEach((path) => {
            if (path.node.params) {
                path.node.params.forEach((param) => {
                    if (param.type === 'Identifier' &&
                        param.typeAnnotation &&
                        param.typeAnnotation.typeAnnotation &&
                        param.typeAnnotation.typeAnnotation.type === 'TSTypeReference' &&
                        param.typeAnnotation.typeAnnotation.typeName &&
                        param.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
                        param.typeAnnotation.typeAnnotation.typeName.name === interfaceName) {
                        param.typeAnnotation.typeAnnotation.typeName.name = newName;
                    }
                });
            }
        });
        rootSource.find(j.Function).forEach((path) => {
            if (path.node.returnType &&
                path.node.returnType.typeAnnotation &&
                path.node.returnType.typeAnnotation.type === 'TSTypeReference' &&
                path.node.returnType.typeAnnotation.typeName &&
                path.node.returnType.typeAnnotation.typeName.type === 'Identifier' &&
                path.node.returnType.typeAnnotation.typeName.name === interfaceName) {
                path.node.returnType.typeAnnotation.typeName.name = newName;
            }
        });
        rootSource.find(j.TSAsExpression).forEach((path) => {
            if (path.node.typeAnnotation &&
                path.node.typeAnnotation.type === 'TSTypeReference' &&
                path.node.typeAnnotation.typeName &&
                path.node.typeAnnotation.typeName.type === 'Identifier' &&
                path.node.typeAnnotation.typeName.name === interfaceName) {
                path.node.typeAnnotation.typeName.name = newName;
            }
        });
    }
};
exports.tsInterfaceTransformer = tsInterfaceTransformer;
