// @progress/kendo-angular-label/esm2022/floating-label/floating-label.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_FLOATINGLABEL } from '../directives';
import * as i0 from "@angular/core";
import * as i1 from "./floating-label.component";
import * as i2 from "../localization/custom-messages.component";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) definition for the FloatingLabel component.
 *
 * @example
 * ```typescript
 * import { FloatingLabelModule } from '@progress/kendo-angular-label';
 *
 * @NgModule({
 *   imports:      [BrowserModule, FloatingLabelModule],
 *   declarations: [AppComponent],
 *   bootstrap:    [AppComponent]
 * })
 * export class AppModule { }
 * ```
 */
export class FloatingLabelModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: FloatingLabelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: FloatingLabelModule, imports: [i1.FloatingLabelComponent, i2.CustomMessagesComponent], exports: [i1.FloatingLabelComponent, i2.CustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: FloatingLabelModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: FloatingLabelModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_FLOATINGLABEL],
                    imports: [...KENDO_FLOATINGLABEL]
                }]
        }] });
