// @progress/kendo-angular-grid/esm2022/pdf/pdf-margin.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component } from '@angular/core';
import { PDFExportMarginComponent } from '@progress/kendo-angular-pdf-export';
import * as i0 from "@angular/core";
/**
 * Represents the Kendo UI PDFMargin component for Angular.
 * Use this component to set the PDF page margins.
 * Supported units are:
 * - `"mm"`
 * - `"cm"`
 * - `"in"`
 * - `"pt"` (default)
 * When you provide only a number, the units default to points (`pt`).
 *
 * @example
 * ```html
 * <kendo-grid>
 *   <kendo-grid-pdf [allPages]="true" [delay]="100">
 *     <kendo-grid-pdf-margin top="1cm" bottom="1cm" left="1cm" right="1cm"></kendo-grid-pdf-margin>
 *   </kendo-grid-pdf>
 * </kendo-grid>
 * ```
 */
export class PDFMarginComponent extends PDFExportMarginComponent {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: PDFMarginComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.14", type: PDFMarginComponent, isStandalone: true, selector: "kendo-grid-pdf-margin", usesInheritance: true, ngImport: i0, template: '', isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: PDFMarginComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'kendo-grid-pdf-margin',
                    template: '',
                    standalone: true
                }]
        }] });
