// @progress/kendo-popup-common/dist/npm/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collision = exports.AlignPoint = exports.getWindowViewPort = exports.getWindow = exports.getDocumentElement = exports.positionElement = exports.utils = exports.domUtils = exports.alignElement = exports.zIndex = exports.siblings = exports.siblingContainer = exports.scrollPosition = exports.restrictToView = exports.removeScroll = exports.positionWithScroll = exports.position = exports.parentScrollPosition = exports.parents = exports.offset = exports.offsetParent = exports.isBodyOffset = exports.boundingOffset = exports.applyLocationOffset = exports.addScroll = exports.align = void 0;
var align_1 = require("./align");
Object.defineProperty(exports, "align", { enumerable: true, get: function () { return align_1.align; } });
var add_scroll_1 = require("./add-scroll");
Object.defineProperty(exports, "addScroll", { enumerable: true, get: function () { return add_scroll_1.addScroll; } });
var apply_location_offset_1 = require("./apply-location-offset");
Object.defineProperty(exports, "applyLocationOffset", { enumerable: true, get: function () { return apply_location_offset_1.applyLocationOffset; } });
var bounding_offset_1 = require("./bounding-offset");
Object.defineProperty(exports, "boundingOffset", { enumerable: true, get: function () { return bounding_offset_1.boundingOffset; } });
var is_body_offset_1 = require("./is-body-offset");
Object.defineProperty(exports, "isBodyOffset", { enumerable: true, get: function () { return is_body_offset_1.isBodyOffset; } });
var offset_parent_1 = require("./offset-parent");
Object.defineProperty(exports, "offsetParent", { enumerable: true, get: function () { return offset_parent_1.offsetParent; } });
var offset_1 = require("./offset");
Object.defineProperty(exports, "offset", { enumerable: true, get: function () { return offset_1.offset; } });
var parents_1 = require("./parents");
Object.defineProperty(exports, "parents", { enumerable: true, get: function () { return parents_1.parents; } });
var parent_scroll_position_1 = require("./parent-scroll-position");
Object.defineProperty(exports, "parentScrollPosition", { enumerable: true, get: function () { return parent_scroll_position_1.parentScrollPosition; } });
var position_1 = require("./position");
Object.defineProperty(exports, "position", { enumerable: true, get: function () { return position_1.position; } });
var position_with_scroll_1 = require("./position-with-scroll");
Object.defineProperty(exports, "positionWithScroll", { enumerable: true, get: function () { return position_with_scroll_1.positionWithScroll; } });
var remove_scroll_1 = require("./remove-scroll");
Object.defineProperty(exports, "removeScroll", { enumerable: true, get: function () { return remove_scroll_1.removeScroll; } });
var restrict_to_view_1 = require("./restrict-to-view");
Object.defineProperty(exports, "restrictToView", { enumerable: true, get: function () { return restrict_to_view_1.restrictToView; } });
var scroll_position_1 = require("./scroll-position");
Object.defineProperty(exports, "scrollPosition", { enumerable: true, get: function () { return scroll_position_1.scrollPosition; } });
var sibling_container_1 = require("./sibling-container");
Object.defineProperty(exports, "siblingContainer", { enumerable: true, get: function () { return sibling_container_1.siblingContainer; } });
var siblings_1 = require("./siblings");
Object.defineProperty(exports, "siblings", { enumerable: true, get: function () { return siblings_1.siblings; } });
var z_index_1 = require("./z-index");
Object.defineProperty(exports, "zIndex", { enumerable: true, get: function () { return z_index_1.zIndex; } });
var align_element_1 = require("./align-element");
Object.defineProperty(exports, "alignElement", { enumerable: true, get: function () { return align_element_1.alignElement; } });
var dom_utils_1 = require("./dom-utils");
Object.defineProperty(exports, "domUtils", { enumerable: true, get: function () { return dom_utils_1.domUtils; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "utils", { enumerable: true, get: function () { return utils_1.utils; } });
var position_element_1 = require("./position-element");
Object.defineProperty(exports, "positionElement", { enumerable: true, get: function () { return position_element_1.positionElement; } });
var document_1 = require("./document");
Object.defineProperty(exports, "getDocumentElement", { enumerable: true, get: function () { return document_1.getDocumentElement; } });
var window_1 = require("./window");
Object.defineProperty(exports, "getWindow", { enumerable: true, get: function () { return window_1.getWindow; } });
var window_viewport_1 = require("./window-viewport");
Object.defineProperty(exports, "getWindowViewPort", { enumerable: true, get: function () { return window_viewport_1.windowViewport; } });
var align_point_1 = require("./align-point");
Object.defineProperty(exports, "AlignPoint", { enumerable: true, get: function () { return align_point_1.AlignPoint; } });
var collision_1 = require("./collision");
Object.defineProperty(exports, "Collision", { enumerable: true, get: function () { return collision_1.Collision; } });
