// @progress/kendo-angular-layout/esm2022/drawer/template-directives/footer-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Optional, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Defines a template that specifies the footer content of the Drawer.
 * To define the template, nest an `<ng-template>` tag with the `kendoDrawerFooterTemplate` directive inside the `<kendo-drawer>` tag.
 *
 * @example
 * ```html
 * <kendo-drawer>
 *   <ng-template kendoDrawerFooterTemplate>
 *     <p>© 2023 Company Name</p>
 *   </ng-template>
 * </kendo-drawer>
 * ```
 */
export class DrawerFooterTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: DrawerFooterTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: DrawerFooterTemplateDirective, isStandalone: true, selector: "[kendoDrawerFooterTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: DrawerFooterTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoDrawerFooterTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }] });
