// @progress/kendo-angular-grid/esm2022/row-reordering/drag-hint-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef, Optional } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * The directive allows customizing the drag hint in the Grid row reorder column.
 * Place an `<ng-template>` tag with the `kendoGridRowDragHintTemplate` directive inside the `<kendo-grid-rowreorder-column>` tag to set the hint template.
 *
 * The template context is the current data item. You also get the data row index as an extra field. Use it as a template variable with `let-rowIndex="rowIndex"`.
 *
 * @example
 * ```html
 * <kendo-grid [kendoGridBinding]="data" [rowReorderable]="true">
 *    <kendo-grid-rowreorder-column>
 *      <ng-template kendoGridRowDragHintTemplate>
 *        <span>Custom hint</span>
 *      </ng-template>
 *    </kendo-grid-rowreorder-column>
 *   <kendo-grid-column field="Field1"></kendo-grid-column>
 *   <kendo-grid-column field="Field2"></kendo-grid-column>
 * </kendo-grid>
 * ```
 */
export class RowDragHintTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: RowDragHintTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: RowDragHintTemplateDirective, isStandalone: true, selector: "[kendoGridRowDragHintTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: RowDragHintTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoGridRowDragHintTemplate]',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }] });
