// @progress/kendo-angular-progressbar/esm2022/circularprogressbar/center-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Represents a template that allows you to customize the content in the center of the `<kendo-circularprogessbar>` component. To define the template, nest an `<ng-template>` tag with the `kendoCircularProgressbarCenterTemplate` directive inside a `<kendo-circularprogessbar>` component.
 * ([see example]({% slug center_template_circularprogressbar %})).
 *
 * @example
 * ```html
 * <kendo-circularprogressbar [value]="75">
 *   <ng-template kendoCircularProgressbarCenterTemplate>
 *     <span class="custom-center-text">75%</span>
 *   </ng-template>
 * </kendo-circularprogressbar>
 * ```
 */
export class CircularProgressbarCenterTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CircularProgressbarCenterTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: CircularProgressbarCenterTemplateDirective, isStandalone: true, selector: "[kendoCircularProgressbarCenterTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: CircularProgressbarCenterTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoCircularProgressbarCenterTemplate]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
