// @progress/kendo-angular-layout/esm2022/tabstrip/directives/tab-title.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Represents the title template of the Kendo UI TabStrip.
 * To define the template, nest an `<ng-template>` tag with the `kendoTabTitle` directive inside the component tag.
 * The `kendoTabTitle` directive overrides the TabStripTab [`title`]({% slug api_layout_tabstriptabcomponent %}#toc-title) option.
 *
 * @example
 * ```html
 * <kendo-tabstrip>
 *   <kendo-tabstrip-tab>
 *     <ng-template kendoTabTitle>
 *       <span class="custom-title">My Custom Title</span>
 *     </ng-template>
 *   </kendo-tabstrip-tab>
 * </kendo-tabstrip>
 * ```
 */
export class TabTitleDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TabTitleDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: TabTitleDirective, isStandalone: true, selector: "[kendoTabTitle]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TabTitleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoTabTitle]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
