// @progress/kendo-angular-layout/esm2022/tabstrip/directives/tab-content.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Represents the content template of the Kendo UI TabStrip.
 * To define the template, nest an `<ng-template>` tag with the `kendoTabContent` directive inside the component tag.
 *
 * @example
 * ```html
 * <kendo-tabstrip>
 *   <kendo-tabstrip-tab title="First Tab">
 *     <ng-template kendoTabContent>
 *       <div class="tab-content">
 *         <h2>Welcome</h2>
 *         <p>This is the content of the first tab.</p>
 *       </div>
 *     </ng-template>
 *   </kendo-tabstrip-tab>
 * </kendo-tabstrip>
 * ```
 */
export class TabContentDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TabContentDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: TabContentDirective, isStandalone: true, selector: "[kendoTabContent]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: TabContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoTabContent]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
