// @progress/kendo-angular-conversational-ui/esm2022/inline-ai-prompt/localization/custom-messages.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
import * as i0 from "@angular/core";
import * as i1 from "@progress/kendo-angular-l10n";
/**
 * Provides a component for customizing messages in the Inline AI Prompt. [See example](slug:globalization_chat#custom-messages).
 *
 * @example
 * ```html
 * <kendo-inlineaiprompt-messages
 *   commandsButtonTitle="Commands Menu"
 *   generateButtonTitle="Generate">
 * </kendo-inlineaiprompt-messages>
 * ```
 */
export class InlineAIPromptCustomMessagesComponent extends Messages {
    service;
    constructor(service) {
        super();
        this.service = service;
    }
    get override() {
        return true;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: InlineAIPromptCustomMessagesComponent, deps: [{ token: i1.LocalizationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.14", type: InlineAIPromptCustomMessagesComponent, isStandalone: true, selector: "kendo-inlineaiprompt-messages", providers: [
            {
                provide: Messages,
                useExisting: forwardRef(() => InlineAIPromptCustomMessagesComponent)
            }
        ], usesInheritance: true, ngImport: i0, template: ``, isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: InlineAIPromptCustomMessagesComponent, decorators: [{
            type: Component,
            args: [{
                    providers: [
                        {
                            provide: Messages,
                            useExisting: forwardRef(() => InlineAIPromptCustomMessagesComponent)
                        }
                    ],
                    selector: 'kendo-inlineaiprompt-messages',
                    template: ``,
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i1.LocalizationService }] });
