// @progress/kendo-angular-buttons/esm2022/speechtotextbutton/speechtotextbutton.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_SPEECHTOTEXTBUTTON } from '../directives';
import { IconsService } from '@progress/kendo-angular-icons';
import * as i0 from "@angular/core";
import * as i1 from "./speechtotextbutton.component";
// IMPORTANT: NgModule export kept for backwards compatibility.
/**
 * Represents the exported package module.
 *
 * Required for adding SpeechToTextButton features in NgModule-based Angular applications.
 *
 * The package exports:
 * - `SpeechToTextButtonComponent`&mdash;The SpeechToTextButton component class.
 */
export class SpeechToTextButtonModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: SpeechToTextButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.14", ngImport: i0, type: SpeechToTextButtonModule, imports: [i1.SpeechToTextButtonComponent], exports: [i1.SpeechToTextButtonComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: SpeechToTextButtonModule, providers: [IconsService], imports: [KENDO_SPEECHTOTEXTBUTTON] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: SpeechToTextButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_SPEECHTOTEXTBUTTON],
                    imports: [...KENDO_SPEECHTOTEXTBUTTON],
                    providers: [IconsService]
                }]
        }] });
