// @progress/kendo-angular-grid/esm2022/navigation/navigation-metadata.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
/**
 * @hidden
 */
export class NavigationMetadata {
    dataRows;
    headerRows;
    isVirtual;
    hasPager;
    hasDetailTemplate;
    gridElement;
    virtualColumns;
    columns;
    footerRow;
    isStacked;
    get maxLogicalRowIndex() {
        const dataRows = this.hasDetailTemplate ? this.dataRows * 2 : this.dataRows;
        return this.headerRows + dataRows + this.footerRow - 1;
    }
    constructor(dataRows, headerRows, isVirtual, hasPager, hasDetailTemplate, gridElement, virtualColumns, columns, footerRow, isStacked) {
        this.dataRows = dataRows;
        this.headerRows = headerRows;
        this.isVirtual = isVirtual;
        this.hasPager = hasPager;
        this.hasDetailTemplate = hasDetailTemplate;
        this.gridElement = gridElement;
        this.virtualColumns = virtualColumns;
        this.columns = columns;
        this.footerRow = footerRow;
        this.isStacked = isStacked;
    }
}
