// @progress/kendo-angular-toolbar/esm2022/tools/toolbar-splitbutton.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, forwardRef, ViewChild, Input, ElementRef, Output, EventEmitter, ViewChildren, QueryList, isDevMode } from '@angular/core';
import { SplitButtonComponent } from '@progress/kendo-angular-buttons';
import { ToolBarToolComponent } from './toolbar-tool.component';
import { getValueForLocation, makePeeker, getIndexOfFocused, getPrevKey, getNextKey, seekFocusedIndex, areEqual } from '../util';
import { caretAltDownIcon } from '@progress/kendo-svg-icons';
import { IconWrapperComponent } from '@progress/kendo-angular-icons';
import { NgClass, NgIf, NgFor } from '@angular/common';
import { ToolBarComponent } from '../toolbar.component';
import { normalizeNumpadKeys } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "../toolbar.component";
/**
 * Represents the [Kendo UI ToolBar SplitButton for Angular](slug:controltypes_toolbar#split-buttons).
 *
 * Use the ToolBar SplitButton to create a split button with a main button and a drop-down list of actions in the ToolBar.
 *
 * @example
 * ```html
 * <kendo-toolbar>
 *      <kendo-toolbar-splitbutton text="Paste" [data]="data">
 *      </kendo-toolbar-splitbutton>
 * </kendo-toolbar>
 * ```
 */
export class ToolBarSplitButtonComponent extends ToolBarToolComponent {
    host;
    // showText and showIcon showIcon should be declared first
    /**
     * Controls the button text visibility.
     * @default 'always'
     */
    set showText(value) {
        this._showText = value;
        this.setTextDisplayMode();
    }
    get showText() {
        return this._showText;
    }
    /**
     * Controls the button icon visibility.
     * @default 'always'
     */
    set showIcon(value) {
        this._showIcon = value;
        this.setTextDisplayMode();
    }
    get showIcon() {
        return this._showIcon;
    }
    /**
     * Sets the text of the SplitButton ([see example](slug:controltypes_toolbar#toc-split-buttons)).
     */
    set text(text) {
        this._text = text;
        this.setTextDisplayMode();
    }
    get text() {
        return this._text;
    }
    /**
     * Sets the icon rendered next to the button text
     * ([see example](slug:controltypes_toolbar#toc-split-buttons)).
     */
    set icon(icon) {
        this.toolbarOptions.icon = getValueForLocation(icon, this.showIcon, false);
        this.overflowOptions.icon = getValueForLocation(icon, this.showIcon, true);
    }
    /**
     * Sets the `SVGIcon` rendered in the main button.
     * Accepts an [existing Kendo SVG icon](slug:svgicon_list) or a custom one.
     */
    set svgIcon(icon) {
        const isIconSet = this.toolbarOptions.icon || this.overflowOptions.icon;
        const isIconClassSet = this.toolbarOptions.iconClass || this.overflowOptions.iconClass;
        if (isDevMode() &&
            icon &&
            isIconSet &&
            isIconClassSet) {
            throw new Error('Setting both icon/svgIcon and iconClass options at the same time is not supported.');
        }
        this.toolbarOptions.svgIcon = getValueForLocation(icon, this.showIcon, false);
        this.overflowOptions.svgIcon = getValueForLocation(icon, this.showIcon, true);
    }
    /**
     * Sets a custom CSS class icon rendered next to the button text.
     */
    set iconClass(iconClass) {
        this.toolbarOptions.iconClass = getValueForLocation(iconClass, this.showIcon, false);
        this.overflowOptions.iconClass = getValueForLocation(iconClass, this.showIcon, true);
    }
    /**
     * Sets a URL for the image displayed next to the button text.
     */
    set imageUrl(imageUrl) {
        this.toolbarOptions.imageUrl = getValueForLocation(imageUrl, this.showIcon, false);
        this.overflowOptions.imageUrl = getValueForLocation(imageUrl, this.showIcon, true);
    }
    /**
     * When `true`, disables a SplitButton item.
     */
    disabled;
    /**
     * Configures the popup of the SplitButton.
     * Accepts a `PopupSettings` object that allows you to customize the popup behavior and appearance.

     */
    set popupSettings(value) {
        this._popupSettings = value;
    }
    get popupSettings() {
        if (!this._popupSettings) {
            this._popupSettings = { animate: true, popupClass: '' };
        }
        return this._popupSettings;
    }
    /**
     * Sets the fill mode for the button.
     * The fill mode represents the background and border styles.
     * @default 'solid'
     */
    fillMode = 'solid';
    /**
     * Sets the predefined theme color for the button.
     * The theme color applies to the background, border, and text.
     * @default 'base'
     */
    themeColor = 'base';
    /**
     * @hidden
     */
    set look(look) {
        if (look) {
            this.fillMode = look === 'default' ? 'solid' : look;
        }
    }
    /**
     * Sets the CSS classes for the main button.
     * Accepts values supported by [`ngClass`](link:site.data.urls.angular['ngclassapi']).
     */
    buttonClass;
    /**
     * Sets the CSS classes for the arrow button that opens the popup.
     * Accepts values supported by [`ngClass`](link:site.data.urls.angular['ngclassapi']).
     */
    arrowButtonClass;
    /**
     * Sets the name of the [font icon](slug:icons#icons-list) for the arrow button.
     */
    arrowButtonIcon = 'caret-alt-down';
    /**
     * Sets the [`SVGIcon`](slug:api_icons_svgicon) for the arrow button.
     */
    arrowButtonSvgIcon = caretAltDownIcon;
    /**
     * Sets the text field for the button-list popup.
     * @default 'text'
     */
    textField = 'text';
    /**
     * Sets the data of the SplitButton ([see example](slug:controltypes_toolbar#toc-split-buttons)).
     *
     * > Provide the data as an array-like list.
     */
    set data(data) {
        this._data = data || [];
    }
    get data() {
        if (!this._data) {
            this.data = [];
        }
        return this._data;
    }
    /**
     * Fires when the user clicks the main button.
     */
    buttonClick = new EventEmitter();
    /**
     * Fires when the user clicks a drop-down list item.
     * The event data contains the clicked item's data.
     */
    itemClick = new EventEmitter();
    /**
     * Fires when the popup is about to open.
     * This event is preventable. If you cancel the event, the popup stays closed.
     */
    open = new EventEmitter();
    /**
     * Fires when the popup is about to close.
     * This event is preventable. If you cancel the event, the popup stays open.
     */
    close = new EventEmitter();
    toolbarOptions = {
        text: '',
        icon: '',
        iconClass: '',
        svgIcon: null,
        imageUrl: ''
    };
    overflowOptions = {
        text: '',
        icon: '',
        iconClass: '',
        svgIcon: null,
        imageUrl: ''
    };
    ngOnInit() {
        this.setTextDisplayMode();
    }
    ngOnDestroy() {
        if (this.propertyChangeSub) {
            this.propertyChangeSub.unsubscribe();
            this.propertyChangeSub = null;
        }
    }
    get overflowButtons() {
        return [this.overflowMainButton, ...this.overflowListItems.toArray().filter(el => !el.nativeElement.classList.contains('k-disabled'))];
    }
    _data;
    _popupSettings = { animate: true, popupClass: '' };
    focusedIndex = -1;
    _showText;
    _showIcon;
    _text;
    propertyChangeSub;
    getNextKey;
    getPrevKey;
    toolbarSplitButton;
    sectionSplitButton;
    overflowMainButton;
    overflowListItems;
    constructor(host) {
        super();
        this.host = host;
        this.getNextKey = getNextKey();
        this.getPrevKey = getPrevKey();
        this.isBuiltInTool = true;
        this.propertyChangeSub = this.host.propertyChange.subscribe(change => {
            if (change.property === 'showText' || change.property === 'showIcon') {
                this[change.property] = change.value;
            }
        });
    }
    /**
     * @hidden
     */
    onButtonListClick(ev) {
        this.focusedIndex = this.overflowButtons.findIndex(b => b.nativeElement.contains(ev.target));
    }
    /**
     * @hidden
     */
    onMainButtonClick(ev) {
        this.buttonClick.emit(ev);
        this.focusedIndex = 0;
    }
    /**
     * @hidden
     */
    canFocus() {
        return !this.disabled;
    }
    /**
     * @hidden
     */
    get size() {
        return this.host.size;
    }
    /**
     * @hidden
     */
    focus(ev = {}) {
        if (!this.overflows || this.location === 'section') {
            if (!ev.type || ev.type === 'focus' || ev.type === 'keydown') {
                this[`${this.location}SplitButton`].focus();
            }
        }
        else if (this.overflowButtons.length > 0) {
            this.focusedIndex = getIndexOfFocused(this.getPrevKey(), this.getNextKey(), this.overflowButtons.map(ob => ob.nativeElement))(ev);
            this.focusButton(this.focusedIndex, ev);
        }
    }
    /**
     * @hidden
     */
    handleKey(ev) {
        const code = normalizeNumpadKeys(ev);
        if (!this.overflows && (code === this.getPrevKey(this.overflows) || code === this.getNextKey(this.overflows))) {
            return false;
        }
        if (this.overflows) {
            const peekAtIndex = makePeeker(this.overflowButtons);
            const isUnmodified = areEqual(this.focusedIndex);
            this.focusedIndex = seekFocusedIndex(this.getPrevKey(), this.getNextKey(), peekAtIndex)(this.focusedIndex, ev);
            this.focusButton(this.focusedIndex, ev);
            return !isUnmodified(this.focusedIndex);
        }
    }
    /**
     * @hidden
     */
    getText(dataItem) {
        if (dataItem) {
            return this.textField ? dataItem[this.textField] : dataItem.text || dataItem;
        }
    }
    /**
     * @hidden
     */
    handleClick(ev, item, index) {
        this.onButtonListClick(ev);
        const dataItem = this.data[index];
        if (item.click) {
            item.click(dataItem);
        }
        this.itemClick.emit(dataItem);
    }
    focusButton(index, ev) {
        if ((!ev.type || ev.type === 'focus' || ev.type === 'keydown' || ev.type === 'click') && this.location !== 'section') {
            this.overflowButtons[index].nativeElement.focus();
        }
    }
    setTextDisplayMode() {
        this.toolbarOptions.text = this.showText === 'menu' || this.showText === 'never' ? undefined : this.text;
        this.overflowOptions.text = this.showText === 'toolbar' || this.showText === 'never' ? undefined : this.text;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ToolBarSplitButtonComponent, deps: [{ token: i1.ToolBarComponent }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.14", type: ToolBarSplitButtonComponent, isStandalone: true, selector: "kendo-toolbar-splitbutton", inputs: { showText: "showText", showIcon: "showIcon", text: "text", icon: "icon", svgIcon: "svgIcon", iconClass: "iconClass", imageUrl: "imageUrl", disabled: "disabled", popupSettings: "popupSettings", fillMode: "fillMode", themeColor: "themeColor", look: "look", buttonClass: "buttonClass", arrowButtonClass: "arrowButtonClass", arrowButtonIcon: "arrowButtonIcon", arrowButtonSvgIcon: "arrowButtonSvgIcon", textField: "textField", data: "data" }, outputs: { buttonClick: "buttonClick", itemClick: "itemClick", open: "open", close: "close" }, providers: [{ provide: ToolBarToolComponent, useExisting: forwardRef(() => ToolBarSplitButtonComponent) }], viewQueries: [{ propertyName: "toolbarSplitButton", first: true, predicate: ["toolbarSplitButton"], descendants: true }, { propertyName: "sectionSplitButton", first: true, predicate: ["sectionSplitButton"], descendants: true }, { propertyName: "overflowMainButton", first: true, predicate: ["overflowMainButton"], descendants: true, read: ElementRef }, { propertyName: "overflowListItems", predicate: ["listItem"], descendants: true }], exportAs: ["kendoToolBarSplitButton"], usesInheritance: true, ngImport: i0, template: `
        <ng-template #toolbarTemplate>
            <kendo-splitbutton
                #toolbarSplitButton
                class="k-toolbar-split-button"
                [data]="data"
                [text]="toolbarOptions.text"
                [icon]="toolbarOptions.icon"
                [iconClass]="toolbarOptions.iconClass"
                [svgIcon]="toolbarOptions.svgIcon"
                [imageUrl]="toolbarOptions.imageUrl"
                [buttonClass]="buttonClass"
                [arrowButtonClass]="arrowButtonClass"
                [arrowButtonIcon]="arrowButtonIcon"
                [arrowButtonSvgIcon]="arrowButtonSvgIcon"
                [disabled]="disabled"
                [size]="size"
                [tabIndex]="-1"
                [textField]="textField"
                [popupSettings]="popupSettings"
                [fillMode]="fillMode"
                [themeColor]="fillMode ? themeColor : null"
                (buttonClick)="buttonClick.emit($event)"
                (open)="open.emit($event)"
                (close)="close.emit($event)"
                (itemClick)="itemClick.emit($event)"
            ></kendo-splitbutton>
        </ng-template>
        <ng-template #popupTemplate>
            <div #overflowMainButton
                tabindex="-1"
                role="menuitem"
                class="k-item k-menu-item"
                [class.k-disabled]="disabled"
                [ngClass]="buttonClass"
                (click)="onMainButtonClick($event)">
                <span
                    [ngClass]="{'k-link': true, 'k-menu-link': true}"
                >
                    <kendo-icon-wrapper
                        *ngIf="overflowOptions.icon || overflowOptions.iconClass || overflowOptions.svgIcon"
                        [name]="overflowOptions.icon"
                        [customFontClass]="overflowOptions.iconClass"
                        [svgIcon]="overflowOptions.svgIcon"
                    >
                    </kendo-icon-wrapper>
                    <span *ngIf="overflowOptions.text" class="k-menu-link-text">{{overflowOptions.text}}</span>
                </span>
            </div>
            <ng-container *ngFor="let item of data; let i = index">
            <div #listItem
                tabindex="-1"
                role="menuitem"
                class="k-item k-menu-item"
                [class.k-disabled]="disabled || item.disabled"
                (click)="handleClick($event, item, i)">
                <span
                    class="k-link k-menu-link"
                >
                    <kendo-icon-wrapper
                        *ngIf="item.icon || item.iconClass || item.svgIcon"
                        [name]="item.icon"
                        [customFontClass]="item.iconClass"
                        [svgIcon]="item.svgIcon"
                    >
                    </kendo-icon-wrapper>
                    <span *ngIf="getText(item)" class="k-menu-link-text">{{ getText(item) }}</span>
                </span>
            </div>
            </ng-container>
        </ng-template>
        <ng-template #sectionTemplate>
            <kendo-splitbutton
                #sectionSplitButton
                class="k-toolbar-split-button"
                [data]="data"
                [text]="toolbarOptions.text"
                [icon]="toolbarOptions.icon"
                [iconClass]="toolbarOptions.iconClass"
                [svgIcon]="toolbarOptions.svgIcon"
                [imageUrl]="toolbarOptions.imageUrl"
                [buttonClass]="buttonClass"
                [arrowButtonClass]="arrowButtonClass"
                [arrowButtonIcon]="arrowButtonIcon"
                [arrowButtonSvgIcon]="arrowButtonSvgIcon"
                [disabled]="disabled"
                [size]="size"
                [tabIndex]="-1"
                [textField]="textField"
                [popupSettings]="popupSettings"
                [fillMode]="fillMode"
                [themeColor]="fillMode ? themeColor : null"
                (buttonClick)="buttonClick.emit($event)"
                (open)="open.emit($event)"
                (close)="close.emit($event)"
                (itemClick)="itemClick.emit($event)"
            ></kendo-splitbutton>
        </ng-template>
    `, isInline: true, dependencies: [{ kind: "component", type: SplitButtonComponent, selector: "kendo-splitbutton", inputs: ["text", "icon", "svgIcon", "iconClass", "type", "imageUrl", "size", "rounded", "fillMode", "themeColor", "disabled", "popupSettings", "tabIndex", "textField", "data", "arrowButtonClass", "arrowButtonIcon", "arrowButtonSvgIcon", "buttonAttributes"], outputs: ["buttonClick", "itemClick", "focus", "blur", "open", "close"], exportAs: ["kendoSplitButton"] }, { kind: "directive", type: NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: IconWrapperComponent, selector: "kendo-icon-wrapper", inputs: ["name", "svgIcon", "innerCssClass", "customFontClass", "size"], exportAs: ["kendoIconWrapper"] }, { kind: "directive", type: NgFor, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ToolBarSplitButtonComponent, decorators: [{
            type: Component,
            args: [{
                    exportAs: 'kendoToolBarSplitButton',
                    providers: [{ provide: ToolBarToolComponent, useExisting: forwardRef(() => ToolBarSplitButtonComponent) }],
                    selector: 'kendo-toolbar-splitbutton',
                    template: `
        <ng-template #toolbarTemplate>
            <kendo-splitbutton
                #toolbarSplitButton
                class="k-toolbar-split-button"
                [data]="data"
                [text]="toolbarOptions.text"
                [icon]="toolbarOptions.icon"
                [iconClass]="toolbarOptions.iconClass"
                [svgIcon]="toolbarOptions.svgIcon"
                [imageUrl]="toolbarOptions.imageUrl"
                [buttonClass]="buttonClass"
                [arrowButtonClass]="arrowButtonClass"
                [arrowButtonIcon]="arrowButtonIcon"
                [arrowButtonSvgIcon]="arrowButtonSvgIcon"
                [disabled]="disabled"
                [size]="size"
                [tabIndex]="-1"
                [textField]="textField"
                [popupSettings]="popupSettings"
                [fillMode]="fillMode"
                [themeColor]="fillMode ? themeColor : null"
                (buttonClick)="buttonClick.emit($event)"
                (open)="open.emit($event)"
                (close)="close.emit($event)"
                (itemClick)="itemClick.emit($event)"
            ></kendo-splitbutton>
        </ng-template>
        <ng-template #popupTemplate>
            <div #overflowMainButton
                tabindex="-1"
                role="menuitem"
                class="k-item k-menu-item"
                [class.k-disabled]="disabled"
                [ngClass]="buttonClass"
                (click)="onMainButtonClick($event)">
                <span
                    [ngClass]="{'k-link': true, 'k-menu-link': true}"
                >
                    <kendo-icon-wrapper
                        *ngIf="overflowOptions.icon || overflowOptions.iconClass || overflowOptions.svgIcon"
                        [name]="overflowOptions.icon"
                        [customFontClass]="overflowOptions.iconClass"
                        [svgIcon]="overflowOptions.svgIcon"
                    >
                    </kendo-icon-wrapper>
                    <span *ngIf="overflowOptions.text" class="k-menu-link-text">{{overflowOptions.text}}</span>
                </span>
            </div>
            <ng-container *ngFor="let item of data; let i = index">
            <div #listItem
                tabindex="-1"
                role="menuitem"
                class="k-item k-menu-item"
                [class.k-disabled]="disabled || item.disabled"
                (click)="handleClick($event, item, i)">
                <span
                    class="k-link k-menu-link"
                >
                    <kendo-icon-wrapper
                        *ngIf="item.icon || item.iconClass || item.svgIcon"
                        [name]="item.icon"
                        [customFontClass]="item.iconClass"
                        [svgIcon]="item.svgIcon"
                    >
                    </kendo-icon-wrapper>
                    <span *ngIf="getText(item)" class="k-menu-link-text">{{ getText(item) }}</span>
                </span>
            </div>
            </ng-container>
        </ng-template>
        <ng-template #sectionTemplate>
            <kendo-splitbutton
                #sectionSplitButton
                class="k-toolbar-split-button"
                [data]="data"
                [text]="toolbarOptions.text"
                [icon]="toolbarOptions.icon"
                [iconClass]="toolbarOptions.iconClass"
                [svgIcon]="toolbarOptions.svgIcon"
                [imageUrl]="toolbarOptions.imageUrl"
                [buttonClass]="buttonClass"
                [arrowButtonClass]="arrowButtonClass"
                [arrowButtonIcon]="arrowButtonIcon"
                [arrowButtonSvgIcon]="arrowButtonSvgIcon"
                [disabled]="disabled"
                [size]="size"
                [tabIndex]="-1"
                [textField]="textField"
                [popupSettings]="popupSettings"
                [fillMode]="fillMode"
                [themeColor]="fillMode ? themeColor : null"
                (buttonClick)="buttonClick.emit($event)"
                (open)="open.emit($event)"
                (close)="close.emit($event)"
                (itemClick)="itemClick.emit($event)"
            ></kendo-splitbutton>
        </ng-template>
    `,
                    standalone: true,
                    imports: [SplitButtonComponent, NgClass, NgIf, IconWrapperComponent, NgFor]
                }]
        }], ctorParameters: () => [{ type: i1.ToolBarComponent }], propDecorators: { showText: [{
                type: Input
            }], showIcon: [{
                type: Input
            }], text: [{
                type: Input
            }], icon: [{
                type: Input
            }], svgIcon: [{
                type: Input
            }], iconClass: [{
                type: Input
            }], imageUrl: [{
                type: Input
            }], disabled: [{
                type: Input
            }], popupSettings: [{
                type: Input
            }], fillMode: [{
                type: Input
            }], themeColor: [{
                type: Input
            }], look: [{
                type: Input
            }], buttonClass: [{
                type: Input
            }], arrowButtonClass: [{
                type: Input
            }], arrowButtonIcon: [{
                type: Input
            }], arrowButtonSvgIcon: [{
                type: Input
            }], textField: [{
                type: Input
            }], data: [{
                type: Input
            }], buttonClick: [{
                type: Output
            }], itemClick: [{
                type: Output
            }], open: [{
                type: Output
            }], close: [{
                type: Output
            }], toolbarSplitButton: [{
                type: ViewChild,
                args: ['toolbarSplitButton', { static: false }]
            }], sectionSplitButton: [{
                type: ViewChild,
                args: ['sectionSplitButton', { static: false }]
            }], overflowMainButton: [{
                type: ViewChild,
                args: ['overflowMainButton', { read: ElementRef }]
            }], overflowListItems: [{
                type: ViewChildren,
                args: ['listItem']
            }] } });
