// @progress/kendo-angular-progressbar/esm2022/common/localization/custom-messages.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { ProgressBarMessages } from './messages';
import * as i0 from "@angular/core";
import * as i1 from "@progress/kendo-angular-l10n";
/**
 * Custom component messages override default component messages
 * ([see example]({% slug rtl_layout %})).
 *
 * @example
 * ```html
 * <!-- Custom messages for ProgressBar -->
 * <kendo-progressbar [value]="value">
 *     <kendo-progressbar-messages
 *       progressBarLabel="Changed ProgressBar Label"
 *     ></kendo-progressbar-messages>
 * </kendo-progressbar>
 *
 * <!-- Custom messages for ChunkProgressBar -->
 * <kendo-chunkprogressbar [value]="value" [chunkCount]="chunks">
 *     <kendo-progressbar-messages
 *       progressBarLabel="Changed Chunk Label"
 *     ></kendo-progressbar-messages>
 * </kendo-chunkprogressbar>
 *
 * <!-- Custom messages for CircularProgressBar -->
 * <kendo-circularprogressbar [value]="value">
 *     <kendo-progressbar-messages
 *       progressBarLabel="Changed Circular Label"
 *     ></kendo-progressbar-messages>
 * </kendo-circularprogressbar>
 * ```
 */
export class ProgressBarCustomMessagesComponent extends ProgressBarMessages {
    service;
    constructor(service) {
        super();
        this.service = service;
    }
    get override() {
        return true;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ProgressBarCustomMessagesComponent, deps: [{ token: i1.LocalizationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.14", type: ProgressBarCustomMessagesComponent, isStandalone: true, selector: "kendo-progressbar-messages", providers: [
            {
                provide: ProgressBarMessages,
                useExisting: forwardRef(() => ProgressBarCustomMessagesComponent)
            }
        ], usesInheritance: true, ngImport: i0, template: ``, isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ProgressBarCustomMessagesComponent, decorators: [{
            type: Component,
            args: [{
                    providers: [
                        {
                            provide: ProgressBarMessages,
                            useExisting: forwardRef(() => ProgressBarCustomMessagesComponent)
                        }
                    ],
                    selector: 'kendo-progressbar-messages',
                    template: ``,
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i1.LocalizationService }] });
