// @progress/kendo-angular-menu/esm2022/context-menu/context-menu-target-container.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, ElementRef } from '@angular/core';
import { ContextMenuTargetService } from './context-menu-target.service';
import * as i0 from "@angular/core";
import * as i1 from "./context-menu-target.service";
/**
 * Represents a container for the [targets]({% slug api_menu_contextmenutargetdirective %}) of the ContextMenu.
 *
 * @example
 * ```html
 * <div
 *   #target="kendoContextMenuTargetContainer"
 *   kendoContextMenuTargetContainer
 * >
 *   <div kendoContextMenuTarget></div>
 *   <div kendoContextMenuTarget></div>
 * </div>
 * ```
 */
export class ContextMenuTargetContainerDirective {
    targetService;
    /**
     * @hidden
     */
    element;
    /**
     * @hidden
     */
    constructor(elementRef, targetService) {
        this.targetService = targetService;
        if (elementRef) {
            this.element = elementRef.nativeElement;
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ContextMenuTargetContainerDirective, deps: [{ token: i0.ElementRef }, { token: i1.ContextMenuTargetService }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.14", type: ContextMenuTargetContainerDirective, isStandalone: true, selector: "[kendoContextMenuTargetContainer]", providers: [ContextMenuTargetService], exportAs: ["kendoContextMenuTargetContainer"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.14", ngImport: i0, type: ContextMenuTargetContainerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoContextMenuTargetContainer]',
                    exportAs: 'kendoContextMenuTargetContainer',
                    providers: [ContextMenuTargetService],
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i1.ContextMenuTargetService }] });
